// react-native-screen-transitions/lib/commonjs/shared/components/create-transition-aware-component.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createTransitionAwareComponent = createTransitionAwareComponent;
var _react = require("react");
var _reactNativeGestureHandler = require("react-native-gesture-handler");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _useAssociatedStyle = require("../hooks/animation/use-associated-style");
var _useScrollRegistry = require("../hooks/gestures/use-scroll-registry");
var _gestures = require("../providers/gestures.provider");
var _registerBounds = require("../providers/register-bounds.provider");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/** biome-ignore-all lint/style/noNonNullAssertion: <This helper is usually being used inside a transitionable stack> */

function createTransitionAwareComponent(Wrapped, options = {}) {
  const {
    isScrollable = false
  } = options;
  const AnimatedComponent = _reactNativeReanimated.default.createAnimatedComponent(Wrapped);
  const ScrollableInner = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
    const {
      nativeGesture
    } = (0, _gestures.useGestureContext)();
    const {
      scrollHandler,
      onContentSizeChange,
      onLayout
    } = (0, _useScrollRegistry.useScrollRegistry)({
      onScroll: props.onScroll,
      onContentSizeChange: props.onContentSizeChange,
      onLayout: props.onLayout
    });
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeGestureHandler.GestureDetector, {
      gesture: nativeGesture,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(AnimatedComponent, {
        ...props,
        ref: ref,
        onScroll: scrollHandler,
        onContentSizeChange: onContentSizeChange,
        onLayout: onLayout,
        scrollEventThrottle: props.scrollEventThrottle || 16
      })
    });
  });
  const Inner = /*#__PURE__*/(0, _react.forwardRef)((props, _) => {
    const {
      children,
      style,
      sharedBoundTag,
      styleId,
      onPress,
      ...rest
    } = props;
    const animatedRef = (0, _reactNativeReanimated.useAnimatedRef)();
    const {
      associatedStyles
    } = (0, _useAssociatedStyle.useAssociatedStyles)({
      id: sharedBoundTag || styleId,
      style
    });
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_registerBounds.RegisterBoundsProvider, {
      animatedRef: animatedRef,
      style: style,
      onPress: onPress,
      sharedBoundTag: sharedBoundTag,
      children: ({
        captureActiveOnPress,
        handleInitialLayout
      }) => /*#__PURE__*/(0, _jsxRuntime.jsx)(AnimatedComponent, {
        ...rest,
        ref: animatedRef,
        style: [style, associatedStyles],
        onPress: captureActiveOnPress,
        onLayout: (0, _reactNativeReanimated.runOnUI)(handleInitialLayout),
        collapsable: !sharedBoundTag,
        children: children
      })
    });
  });
  if (isScrollable) {
    return /*#__PURE__*/(0, _react.memo)(ScrollableInner);
  }
  return /*#__PURE__*/(0, _react.memo)(Inner);
}
