// react-native-screen-transitions/lib/commonjs/shared/components/root-transition-aware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RootTransitionAware = void 0;
var _react = require("react");
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _constants = require("../constants");
var _useStackPointerEvents = require("../hooks/use-stack-pointer-events");
var _styles = require("../providers/screen/styles.provider");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const RootTransitionAware = exports.RootTransitionAware = /*#__PURE__*/(0, _react.memo)(({
  children
}) => {
  const {
    stylesMap
  } = (0, _styles.useScreenStyles)();
  const pointerEvents = (0, _useStackPointerEvents.useStackPointerEvents)();
  const animatedContentStyle = (0, _reactNativeReanimated.useAnimatedStyle)(() => {
    "worklet";

    return stylesMap.value.contentStyle || _constants.NO_STYLES;
  });
  const animatedOverlayStyle = (0, _reactNativeReanimated.useAnimatedStyle)(() => {
    "worklet";

    return stylesMap.value.overlayStyle || _constants.NO_STYLES;
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: [styles.container],
    pointerEvents: pointerEvents,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeReanimated.default.View, {
      style: [_reactNative.StyleSheet.absoluteFillObject, animatedOverlayStyle],
      pointerEvents: "none"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeReanimated.default.View, {
      style: [styles.content, animatedContentStyle],
      pointerEvents: pointerEvents,
      children: children
    })]
  });
});
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  },
  content: {
    flex: 1
  }
});
