// react-native-screen-transitions/lib/commonjs/component-stack/navigators/create-component-stack-navigator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createComponentStackNavigator = createComponentStackNavigator;
var _native = require("@react-navigation/native");
var React = _interopRequireWildcard(require("react"));
var _stackView = require("../components/stack-view");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * Context to detect if we're inside a ComponentStack.
 * Used to avoid double-wrapping nested ComponentStacks in NavigationIndependentTree.
 */
const ComponentStackContext = /*#__PURE__*/React.createContext(false);
ComponentStackContext.displayName = "ComponentStackContext";
const ComponentStackNavigatorInner = /*#__PURE__*/React.memo(function ComponentStackNavigatorInner({
  id,
  initialRouteName,
  children,
  layout,
  screenListeners,
  screenOptions,
  screenLayout,
  ...rest
}) {
  const {
    state,
    describe,
    descriptors,
    navigation,
    NavigationContent
  } = (0, _native.useNavigationBuilder)(_native.StackRouter, {
    id,
    initialRouteName,
    children,
    layout,
    screenListeners,
    screenOptions,
    screenLayout
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(NavigationContent, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_stackView.StackView, {
      ...rest,
      state: state,
      navigation: navigation,
      descriptors: descriptors,
      describe: describe
    })
  });
});

/**
 * ComponentStackNavigator conditionally wraps in NavigationIndependentTree.
 * - If already inside a ComponentStack (nested), just render the inner navigator
 *   so it participates in the parent's navigation tree and goBack() works.
 * - If top-level, wrap in NavigationIndependentTree + NavigationContainer
 *   to isolate from Expo Router / React Navigation.
 */
function IsolatedComponentStackNavigator(props) {
  const isNested = React.useContext(ComponentStackContext);
  if (isNested) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(ComponentStackNavigatorInner, {
      ...props
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_native.NavigationIndependentTree, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_native.NavigationContainer, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ComponentStackContext.Provider, {
        value: true,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ComponentStackNavigatorInner, {
          ...props
        })
      })
    })
  });
}
function createComponentStackNavigator(config) {
  return (0, _native.createNavigatorFactory)(IsolatedComponentStackNavigator)(config);
}
