// react-native-screen-transitions/lib/commonjs/blank-stack/components/adjusted-screen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdjustedScreen = void 0;
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _reactNativeScreens = require("react-native-screens");
var _useStack = require("../../shared/hooks/navigation/use-stack");
var _layoutAnchor = require("../../shared/providers/layout-anchor.provider");
var _managed = require("../../shared/providers/stack/managed.provider");
var _animation = require("../../shared/stores/animation.store");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
var ScreenActivity = /*#__PURE__*/function (ScreenActivity) {
  ScreenActivity[ScreenActivity["INACTIVE"] = 0] = "INACTIVE";
  ScreenActivity[ScreenActivity["TRANSITIONING_OR_BELOW_TOP"] = 1] = "TRANSITIONING_OR_BELOW_TOP";
  ScreenActivity[ScreenActivity["ON_TOP"] = 2] = "ON_TOP";
  return ScreenActivity;
}(ScreenActivity || {});
const EPSILON = 1e-5;
const POINT_NONE = "none";
const POINT_BOX_NONE = "box-none";
const AnimatedNativeScreen = _reactNativeReanimated.default.createAnimatedComponent(_reactNativeScreens.Screen);
const AdjustedScreen = ({
  routeKey,
  index,
  isPreloaded,
  children,
  freezeOnBlur,
  shouldFreeze
}) => {
  const {
    flags: {
      DISABLE_NATIVE_SCREENS = false
    }
  } = (0, _useStack.useStack)();
  const {
    activeScreensLimit,
    routes
  } = (0, _managed.useManagedStackContext)();
  const routesLength = routes.length;
  const screenRef = (0, _reactNativeReanimated.useAnimatedRef)();
  const sceneProgress = _animation.AnimationStore.getAnimation(routeKey, "progress");
  const sceneClosing = _animation.AnimationStore.getAnimation(routeKey, "closing");
  const screenActivity = (0, _reactNativeReanimated.useSharedValue)(ScreenActivity.TRANSITIONING_OR_BELOW_TOP);
  (0, _reactNativeReanimated.useDerivedValue)(() => {
    if (!sceneProgress) {
      screenActivity.set(ScreenActivity.TRANSITIONING_OR_BELOW_TOP);
      return;
    }
    if (index < routesLength - activeScreensLimit - 1 || isPreloaded) {
      screenActivity.set(ScreenActivity.INACTIVE);
    } else {
      const outputValue = index === routesLength - 1 ? ScreenActivity.ON_TOP : index >= routesLength - activeScreensLimit ? ScreenActivity.TRANSITIONING_OR_BELOW_TOP : ScreenActivity.INACTIVE;
      const v = (0, _reactNativeReanimated.interpolate)(sceneProgress.get(), [0, 1 - EPSILON, 1], [1, 1, outputValue], _reactNativeReanimated.Extrapolation.CLAMP);
      const next = Math.trunc(v) ?? ScreenActivity.TRANSITIONING_OR_BELOW_TOP;
      if (next !== screenActivity.get()) {
        screenActivity.set(next);
      }
    }
  });
  const animatedProps = (0, _reactNativeReanimated.useAnimatedProps)(() => {
    const activity = screenActivity.get();
    if (!DISABLE_NATIVE_SCREENS) {
      return {
        activityState: activity,
        shouldFreeze: activity === ScreenActivity.INACTIVE && shouldFreeze,
        pointerEvents: sceneClosing.get() ? POINT_NONE : POINT_BOX_NONE
      };
    }
    return {
      pointerEvents: sceneClosing.get() ? POINT_NONE : POINT_BOX_NONE
    };
  });
  const AdjustedScreenComponent = !DISABLE_NATIVE_SCREENS ? AnimatedNativeScreen : _reactNativeReanimated.default.View;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(AdjustedScreenComponent, {
    enabled: true,
    ref: screenRef,
    style: _reactNative.StyleSheet.absoluteFill,
    freezeOnBlur: freezeOnBlur,
    animatedProps: animatedProps,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_layoutAnchor.LayoutAnchorProvider, {
      anchorRef: screenRef,
      children: children
    })
  });
};
exports.AdjustedScreen = AdjustedScreen;
