// react-native-screen-transitions/lib/commonjs/blank-stack/components/stack-view.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StackView = void 0;
var _native = require("@react-navigation/native");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _overlay = require("../../shared/components/overlay");
var _screenComposer = require("../../shared/providers/screen/screen-composer");
var _core = require("../../shared/providers/stack/core.provider");
var _managed = require("../../shared/providers/stack/managed.provider");
var _adjustedScreen = require("./adjusted-screen");
var _adjustedScreenContainer = require("./adjusted-screen-container");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function isFabric() {
  return "nativeFabricUIManager" in global;
}
const SceneView = /*#__PURE__*/React.memo(function SceneView({
  descriptor
}) {
  const {
    route,
    navigation,
    render
  } = descriptor;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_native.NavigationContext.Provider, {
    value: navigation,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_native.NavigationRouteContext.Provider, {
      value: route,
      children: [descriptor.options.overlayMode === "screen" && /*#__PURE__*/(0, _jsxRuntime.jsx)(_overlay.Overlay.Screen, {}), render()]
    })
  });
});
const StackView = exports.StackView = (0, _core.withStackCore)({
  TRANSITIONS_ALWAYS_ON: true,
  DISABLE_NATIVE_SCREENS: false
}, (0, _managed.withManagedStack)(({
  descriptors,
  focusedIndex,
  scenes,
  shouldShowFloatOverlay
}) => {
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_react.Fragment, {
    children: [shouldShowFloatOverlay ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_overlay.Overlay.Float, {}) : null, /*#__PURE__*/(0, _jsxRuntime.jsx)(_adjustedScreenContainer.AdjustedScreenContainer, {
      children: scenes.map((scene, sceneIndex) => {
        const descriptor = scene.descriptor;
        const route = scene.route;
        const isFocused = focusedIndex === sceneIndex;
        const isBelowFocused = focusedIndex - 1 === sceneIndex;
        const previousDescriptor = scenes[sceneIndex - 1]?.descriptor ?? undefined;
        const nextDescriptor = scenes[sceneIndex + 1]?.descriptor ?? undefined;
        const isPreloaded = descriptors[route.key] === undefined;

        // On Fabric, when screen is frozen, animated and reanimated values are not updated
        // due to component being unmounted. To avoid this, we don't freeze the previous screen there
        const shouldFreeze = isFabric() ? !isPreloaded && !isFocused && !isBelowFocused : !isPreloaded && !isFocused;
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(_adjustedScreen.AdjustedScreen, {
          isPreloaded: isPreloaded,
          index: sceneIndex,
          routeKey: route.key,
          shouldFreeze: shouldFreeze,
          freezeOnBlur: descriptor.options.freezeOnBlur,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_screenComposer.ScreenComposer, {
            previous: previousDescriptor,
            current: descriptor,
            next: nextDescriptor,
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(SceneView, {
              descriptor: descriptor
            }, route.key)
          })
        }, route.key);
      })
    })]
  });
}));
