// react-native-gesture-image-viewer/lib/module/GestureViewerRegistry.js
"use strict";

import GestureViewerManager from "./GestureViewerManager.js";
class GestureViewerRegistry {
  managers = new Map();
  subscribers = new Map();
  triggers = new Map();
  subscribeToManager(id, callback) {
    if (!this.subscribers.has(id)) {
      this.subscribers.set(id, new Set());
    }
    this.subscribers.get(id)?.add(callback);
    const manager = this.managers.get(id) || null;
    callback(manager);
    return () => {
      const subscribers = this.subscribers.get(id);
      subscribers?.delete(callback);
      if (subscribers && subscribers.size === 0) {
        this.subscribers.delete(id);
      }
    };
  }
  createManager(id) {
    if (this.managers.has(id)) {
      return this.managers.get(id) || null;
    }
    const manager = new GestureViewerManager();
    this.managers.set(id, manager);
    this.notifySubscribers(id, manager);
    return manager;
  }
  getManager(id) {
    return this.managers.get(id) || null;
  }
  deleteManager(id) {
    const manager = this.managers.get(id);
    if (manager) {
      manager.cleanUp();
      this.managers.delete(id);
      this.clearTriggerNode(id);
      this.notifySubscribers(id, null);
    }
  }
  notifySubscribers(id, manager) {
    const listeners = this.subscribers.get(id);
    if (listeners) {
      [...listeners].forEach(callback => callback(manager));
    }
  }
  setTriggerNode(id, node) {
    if (!node) {
      this.clearTriggerNode(id);
      return;
    }
    this.triggers.set(id, node);
  }
  getTriggerNode(id) {
    return this.triggers.get(id) ?? null;
  }
  clearTriggerNode(id) {
    this.triggers.delete(id);
  }
}
export const registry = new GestureViewerRegistry();
