// react-native-screen-transitions/lib/module/blank-stack/components/stack-view.native.js
"use strict";

import { NavigationContext, NavigationRouteContext } from "@react-navigation/native";
import * as React from "react";
import { Fragment } from "react";
import { Overlay } from "../../shared/components/overlay";
import { ScreenComposer } from "../../shared/providers/screen/screen-composer";
import { withStackCore } from "../../shared/providers/stack/core.provider";
import { withManagedStack } from "../../shared/providers/stack/managed.provider";
import { AdjustedScreen } from "./adjusted-screen";
import { AdjustedScreenContainer } from "./adjusted-screen-container";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
function isFabric() {
  return "nativeFabricUIManager" in global;
}
const SceneView = /*#__PURE__*/React.memo(function SceneView({
  descriptor
}) {
  const {
    route,
    navigation,
    render
  } = descriptor;
  return /*#__PURE__*/_jsx(NavigationContext.Provider, {
    value: navigation,
    children: /*#__PURE__*/_jsxs(NavigationRouteContext.Provider, {
      value: route,
      children: [descriptor.options.overlayMode === "screen" && /*#__PURE__*/_jsx(Overlay.Screen, {}), render()]
    })
  });
});
export const StackView = withStackCore({
  TRANSITIONS_ALWAYS_ON: true,
  DISABLE_NATIVE_SCREENS: false
}, withManagedStack(({
  descriptors,
  focusedIndex,
  scenes,
  shouldShowFloatOverlay
}) => {
  return /*#__PURE__*/_jsxs(Fragment, {
    children: [shouldShowFloatOverlay ? /*#__PURE__*/_jsx(Overlay.Float, {}) : null, /*#__PURE__*/_jsx(AdjustedScreenContainer, {
      children: scenes.map((scene, sceneIndex) => {
        const descriptor = scene.descriptor;
        const route = scene.route;
        const isFocused = focusedIndex === sceneIndex;
        const isBelowFocused = focusedIndex - 1 === sceneIndex;
        const previousDescriptor = scenes[sceneIndex - 1]?.descriptor ?? undefined;
        const nextDescriptor = scenes[sceneIndex + 1]?.descriptor ?? undefined;
        const isPreloaded = descriptors[route.key] === undefined;

        // On Fabric, when screen is frozen, animated and reanimated values are not updated
        // due to component being unmounted. To avoid this, we don't freeze the previous screen there
        const shouldFreeze = isFabric() ? !isPreloaded && !isFocused && !isBelowFocused : !isPreloaded && !isFocused;
        return /*#__PURE__*/_jsx(AdjustedScreen, {
          isPreloaded: isPreloaded,
          index: sceneIndex,
          routeKey: route.key,
          shouldFreeze: shouldFreeze,
          freezeOnBlur: descriptor.options.freezeOnBlur,
          children: /*#__PURE__*/_jsx(ScreenComposer, {
            previous: previousDescriptor,
            current: descriptor,
            next: nextDescriptor,
            children: /*#__PURE__*/_jsx(SceneView, {
              descriptor: descriptor
            }, route.key)
          })
        }, route.key);
      })
    })]
  });
}));
