// react-native-screen-transitions/lib/module/shared/stores/gesture.store.js
"use strict";

import { cancelAnimation, makeMutable } from "react-native-reanimated";
const store = {};
function ensure(routeKey) {
  let bag = store[routeKey];
  if (!bag) {
    bag = {
      x: makeMutable(0),
      y: makeMutable(0),
      normalizedX: makeMutable(0),
      normalizedY: makeMutable(0),
      isDismissing: makeMutable(0),
      isDragging: makeMutable(0),
      direction: makeMutable(null)
    };
    store[routeKey] = bag;
  }
  return bag;
}
function getGesture(routeKey, gestureKey) {
  return ensure(routeKey)[gestureKey];
}
function getRouteGestures(routeKey) {
  return ensure(routeKey);
}
function clear(routeKey) {
  const bag = store[routeKey];
  if (bag) {
    cancelAnimation(bag.x);
    cancelAnimation(bag.y);
    cancelAnimation(bag.normalizedX);
    cancelAnimation(bag.normalizedY);
    cancelAnimation(bag.isDismissing);
    cancelAnimation(bag.isDragging);
    cancelAnimation(bag.direction);
  }
  delete store[routeKey];
}
export const GestureStore = {
  getGesture,
  getRouteGestures,
  clear
};
