// react-native-screen-transitions/lib/commonjs/shared/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TRUE = exports.NO_STYLES = exports.MASK_STYLE_ID = exports.IS_WEB = exports.GESTURE_VELOCITY_IMPACT = exports.FULLSCREEN_DIMENSIONS = exports.FALSE = exports.EXIT_RANGE = exports.ENTER_RANGE = exports.EMPTY_BOUND_HELPER_RESULT_RAW = exports.EMPTY_BOUND_HELPER_RESULT = exports.DEFAULT_SCREEN_TRANSITION_STATE = exports.DEFAULT_GESTURE_DRIVES_PROGRESS = exports.DEFAULT_GESTURE_DIRECTION = exports.DEFAULT_GESTURE_ACTIVATION_AREA = exports.CONTAINER_STYLE_ID = void 0;
var _reactNative = require("react-native");
/**
 * Masked view integration
 */
const MASK_STYLE_ID = exports.MASK_STYLE_ID = "_ROOT_MASKED";
const CONTAINER_STYLE_ID = exports.CONTAINER_STYLE_ID = "_ROOT_CONTAINER";

/**
 * Styles
 */
const NO_STYLES = exports.NO_STYLES = Object.freeze({});

/**
 * Default screen transition state
 */
const DEFAULT_SCREEN_TRANSITION_STATE = exports.DEFAULT_SCREEN_TRANSITION_STATE = Object.freeze({
  progress: 0,
  closing: 0,
  animating: 0,
  entering: 1,
  gesture: {
    x: 0,
    y: 0,
    normalizedX: 0,
    normalizedY: 0,
    isDismissing: 0,
    isDragging: 0,
    direction: null
  },
  route: {}
});

/**
 * Bounds API Defaults
 */
const EMPTY_BOUND_HELPER_RESULT = exports.EMPTY_BOUND_HELPER_RESULT = Object.freeze({});
const EMPTY_BOUND_HELPER_RESULT_RAW = exports.EMPTY_BOUND_HELPER_RESULT_RAW = Object.freeze({
  scaleX: 1,
  scaleY: 1,
  scale: 1,
  translateX: 0,
  translateY: 0,
  width: 0,
  height: 0
});
const ENTER_RANGE = exports.ENTER_RANGE = [0, 1];
const EXIT_RANGE = exports.EXIT_RANGE = [1, 2];
const FULLSCREEN_DIMENSIONS = dimensions => {
  "worklet";

  return {
    x: 0,
    y: 0,
    pageX: 0,
    pageY: 0,
    width: dimensions.width,
    height: dimensions.height
  };
};

/**
 * Default gesture config
 */
exports.FULLSCREEN_DIMENSIONS = FULLSCREEN_DIMENSIONS;
const GESTURE_VELOCITY_IMPACT = exports.GESTURE_VELOCITY_IMPACT = 0.3;
const DEFAULT_GESTURE_DIRECTION = exports.DEFAULT_GESTURE_DIRECTION = "horizontal";
const DEFAULT_GESTURE_DRIVES_PROGRESS = exports.DEFAULT_GESTURE_DRIVES_PROGRESS = true;
const DEFAULT_GESTURE_ACTIVATION_AREA = exports.DEFAULT_GESTURE_ACTIVATION_AREA = "screen";
const IS_WEB = exports.IS_WEB = _reactNative.Platform.OS === "web";
const TRUE = exports.TRUE = 1;
const FALSE = exports.FALSE = 0;
