// react-native-screen-transitions/lib/commonjs/shared/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
Object.defineProperty(exports, "useHistory", {
  enumerable: true,
  get: function () {
    return _useHistory.useHistory;
  }
});
Object.defineProperty(exports, "useScreenAnimation", {
  enumerable: true,
  get: function () {
    return _useScreenAnimation.useScreenAnimation;
  }
});
Object.defineProperty(exports, "useScreenGesture", {
  enumerable: true,
  get: function () {
    return _useScreenGesture.useScreenGesture;
  }
});
Object.defineProperty(exports, "useScreenState", {
  enumerable: true,
  get: function () {
    return _useScreenState.useScreenState;
  }
});
var _reactNative = require("react-native");
var _createTransitionAwareComponent = require("./components/create-transition-aware-component");
var _maskedView = _interopRequireDefault(require("./components/integrations/masked-view"));
var _configs = require("./configs");
var _useScreenAnimation = require("./hooks/animation/use-screen-animation");
var _useScreenGesture = require("./hooks/gestures/use-screen-gesture");
var _useHistory = require("./hooks/navigation/use-history");
var _useScreenState = require("./hooks/navigation/use-screen-state");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = {
  createTransitionAwareComponent: _createTransitionAwareComponent.createTransitionAwareComponent,
  View: (0, _createTransitionAwareComponent.createTransitionAwareComponent)(_reactNative.View),
  Pressable: (0, _createTransitionAwareComponent.createTransitionAwareComponent)(_reactNative.Pressable),
  ScrollView: (0, _createTransitionAwareComponent.createTransitionAwareComponent)(_reactNative.ScrollView, {
    isScrollable: true
  }),
  FlatList: (0, _createTransitionAwareComponent.createTransitionAwareComponent)(_reactNative.FlatList, {
    isScrollable: true
  }),
  MaskedView: _maskedView.default,
  Presets: _configs.Presets,
  Specs: _configs.Specs
};
