// react-native-screen-transitions/lib/commonjs/shared/stores/gesture.store.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GestureStore = void 0;
var _reactNativeReanimated = require("react-native-reanimated");
const store = {};
function ensure(routeKey) {
  let bag = store[routeKey];
  if (!bag) {
    bag = {
      x: (0, _reactNativeReanimated.makeMutable)(0),
      y: (0, _reactNativeReanimated.makeMutable)(0),
      normalizedX: (0, _reactNativeReanimated.makeMutable)(0),
      normalizedY: (0, _reactNativeReanimated.makeMutable)(0),
      isDismissing: (0, _reactNativeReanimated.makeMutable)(0),
      isDragging: (0, _reactNativeReanimated.makeMutable)(0),
      direction: (0, _reactNativeReanimated.makeMutable)(null)
    };
    store[routeKey] = bag;
  }
  return bag;
}
function getGesture(routeKey, gestureKey) {
  return ensure(routeKey)[gestureKey];
}
function getRouteGestures(routeKey) {
  return ensure(routeKey);
}
function clear(routeKey) {
  const bag = store[routeKey];
  if (bag) {
    (0, _reactNativeReanimated.cancelAnimation)(bag.x);
    (0, _reactNativeReanimated.cancelAnimation)(bag.y);
    (0, _reactNativeReanimated.cancelAnimation)(bag.normalizedX);
    (0, _reactNativeReanimated.cancelAnimation)(bag.normalizedY);
    (0, _reactNativeReanimated.cancelAnimation)(bag.isDismissing);
    (0, _reactNativeReanimated.cancelAnimation)(bag.isDragging);
    (0, _reactNativeReanimated.cancelAnimation)(bag.direction);
  }
  delete store[routeKey];
}
const GestureStore = exports.GestureStore = {
  getGesture,
  getRouteGestures,
  clear
};
