// @launchdarkly/react-native-client-sdk/dist/src/platform/PlatformInfo.js
import { name, version } from '../../package.json';
import { ldApplication, ldDevice } from './autoEnv';
export default class PlatformInfo {
    constructor(_logger, _config) {
        this._logger = _logger;
        this._config = _config;
    }
    platformData() {
        const data = {
            name: 'React Native',
            ld_application: ldApplication,
            ld_device: ldDevice,
        };
        this._logger.debug(`platformData: ${JSON.stringify(data, null, 2)}`);
        return data;
    }
    sdkData() {
        var _a, _b;
        const data = {
            name,
            version,
            userAgentBase: 'ReactNativeClient',
        };
        if ((_a = this._config) === null || _a === void 0 ? void 0 : _a.wrapperName) {
            data.wrapperName = this._config.wrapperName;
        }
        if ((_b = this._config) === null || _b === void 0 ? void 0 : _b.wrapperVersion) {
            data.wrapperVersion = this._config.wrapperVersion;
        }
        this._logger.debug(`sdkData: ${JSON.stringify(data, null, 2)}`);
        return data;
    }
}
