// react-native-screen-transitions/lib/module/shared/components/create-transition-aware-component.js
"use strict";

/** biome-ignore-all lint/style/noNonNullAssertion: <This helper is usually being used inside a transitionable stack> */

import { forwardRef, memo } from "react";
import { GestureDetector } from "react-native-gesture-handler";
import Animated, { runOnUI, useAnimatedRef } from "react-native-reanimated";
import { useAssociatedStyles } from "../hooks/animation/use-associated-style";
import { useScrollRegistry } from "../hooks/gestures/use-scroll-registry";
import { useGestureContext } from "../providers/gestures.provider";
import { RegisterBoundsProvider } from "../providers/register-bounds.provider";
import { jsx as _jsx } from "react/jsx-runtime";
export function createTransitionAwareComponent(Wrapped, options = {}) {
  const {
    isScrollable = false
  } = options;
  const AnimatedComponent = Animated.createAnimatedComponent(Wrapped);
  const ScrollableInner = /*#__PURE__*/forwardRef((props, ref) => {
    const {
      nativeGesture
    } = useGestureContext();
    const {
      scrollHandler,
      onContentSizeChange,
      onLayout
    } = useScrollRegistry({
      onScroll: props.onScroll,
      onContentSizeChange: props.onContentSizeChange,
      onLayout: props.onLayout
    });
    return /*#__PURE__*/_jsx(GestureDetector, {
      gesture: nativeGesture,
      children: /*#__PURE__*/_jsx(AnimatedComponent, {
        ...props,
        ref: ref,
        onScroll: scrollHandler,
        onContentSizeChange: onContentSizeChange,
        onLayout: onLayout,
        scrollEventThrottle: props.scrollEventThrottle || 16
      })
    });
  });
  const Inner = /*#__PURE__*/forwardRef((props, _) => {
    const {
      children,
      style,
      sharedBoundTag,
      styleId,
      onPress,
      ...rest
    } = props;
    const animatedRef = useAnimatedRef();
    const {
      associatedStyles
    } = useAssociatedStyles({
      id: sharedBoundTag || styleId,
      style
    });
    return /*#__PURE__*/_jsx(RegisterBoundsProvider, {
      animatedRef: animatedRef,
      style: style,
      onPress: onPress,
      sharedBoundTag: sharedBoundTag,
      children: ({
        captureActiveOnPress,
        handleInitialLayout
      }) => /*#__PURE__*/_jsx(AnimatedComponent, {
        ...rest,
        ref: animatedRef,
        style: [style, associatedStyles],
        onPress: captureActiveOnPress,
        onLayout: runOnUI(handleInitialLayout),
        collapsable: !sharedBoundTag,
        children: children
      })
    });
  });
  if (isScrollable) {
    return /*#__PURE__*/memo(ScrollableInner);
  }
  return /*#__PURE__*/memo(Inner);
}
