// react-native-screen-transitions/lib/module/shared/components/root-transition-aware.js
"use strict";

import { memo } from "react";
import { StyleSheet, View } from "react-native";
import Animated, { useAnimatedStyle } from "react-native-reanimated";
import { NO_STYLES } from "../constants";
import { useStackPointerEvents } from "../hooks/use-stack-pointer-events";
import { useScreenStyles } from "../providers/screen/styles.provider";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
export const RootTransitionAware = /*#__PURE__*/memo(({
  children
}) => {
  const {
    stylesMap
  } = useScreenStyles();
  const pointerEvents = useStackPointerEvents();
  const animatedContentStyle = useAnimatedStyle(() => {
    "worklet";

    return stylesMap.value.contentStyle || NO_STYLES;
  });
  const animatedOverlayStyle = useAnimatedStyle(() => {
    "worklet";

    return stylesMap.value.overlayStyle || NO_STYLES;
  });
  return /*#__PURE__*/_jsxs(View, {
    style: [styles.container],
    pointerEvents: pointerEvents,
    children: [/*#__PURE__*/_jsx(Animated.View, {
      style: [StyleSheet.absoluteFillObject, animatedOverlayStyle],
      pointerEvents: "none"
    }), /*#__PURE__*/_jsx(Animated.View, {
      style: [styles.content, animatedContentStyle],
      pointerEvents: pointerEvents,
      children: children
    })]
  });
});
const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  content: {
    flex: 1
  }
});
