// react-native-screen-transitions/lib/module/shared/providers/stack/core.provider.js
"use strict";

import { SafeAreaProviderCompat } from "@react-navigation/elements";
import { StyleSheet } from "react-native";
import { GestureHandlerRootView } from "react-native-gesture-handler";
import { StackType } from "../../types/stack.types";
import createProvider from "../../utils/create-provider";
import { jsx as _jsx } from "react/jsx-runtime";
const {
  StackCoreProvider: InternalStackCoreProvider,
  useStackCoreContext
} = createProvider("StackCore", {
  guarded: true
})(({
  config,
  children
}) => {
  const {
    TRANSITIONS_ALWAYS_ON = false,
    DISABLE_NATIVE_SCREENS = false,
    STACK_TYPE = StackType.BLANK
  } = config;
  return {
    value: {
      flags: {
        TRANSITIONS_ALWAYS_ON,
        STACK_TYPE,
        DISABLE_NATIVE_SCREENS
      }
    },
    children: /*#__PURE__*/_jsx(GestureHandlerRootView, {
      style: styles.container,
      pointerEvents: STACK_TYPE === StackType.COMPONENT ? "box-none" : undefined,
      children: /*#__PURE__*/_jsx(SafeAreaProviderCompat, {
        children: children
      })
    })
  };
});

/**
 * HOC that wraps a component with the StackCore provider.
 * Just a simple open gate
 */
export function withStackCore(defaultConfig, Component) {
  return function StackCoreWrapper({
    DISABLE_NATIVE_SCREENS,
    TRANSITIONS_ALWAYS_ON,
    STACK_TYPE,
    ...props
  }) {
    // User props first, then defaultConfig overrides (config takes priority)
    const config = {
      ...(DISABLE_NATIVE_SCREENS !== undefined && {
        DISABLE_NATIVE_SCREENS
      }),
      ...(TRANSITIONS_ALWAYS_ON !== undefined && {
        TRANSITIONS_ALWAYS_ON
      }),
      ...(STACK_TYPE !== undefined && {
        STACK_TYPE
      }),
      ...defaultConfig
    };
    return /*#__PURE__*/_jsx(InternalStackCoreProvider, {
      config: config,
      children: /*#__PURE__*/_jsx(Component, {
        ...props
      })
    });
  };
}
const styles = StyleSheet.create({
  container: {
    flex: 1
  }
});
export { useStackCoreContext };
