// react-native-screen-transitions/lib/module/shared/providers/stack/direct.provider.js
"use strict";

import * as React from "react";
import { useEffect, useMemo } from "react";
import { useDerivedValue } from "react-native-reanimated";
import { StackContext } from "../../hooks/navigation/use-stack";
import { AnimationStore } from "../../stores/animation.store";
import { HistoryStore } from "../../stores/history.store";
import { useStackCoreContext } from "./core.provider";
import { jsx as _jsx } from "react/jsx-runtime";
const DirectStackContext = /*#__PURE__*/React.createContext(null);
DirectStackContext.displayName = "DirectStack";
function useDirectStackContext() {
  const context = React.useContext(DirectStackContext);
  if (!context) {
    throw new Error("useDirectStackContext must be used within DirectStackProvider");
  }
  return context;
}

/**
 * Internal hook that computes all lifecycle values.
 */
function useDirectStackValue(props) {
  const {
    state,
    navigation,
    descriptors,
    describe
  } = props;
  const {
    flags
  } = useStackCoreContext();
  const preloadedDescriptors = useMemo(() => {
    return state.preloadedRoutes.reduce((acc, route) => {
      acc[route.key] = acc[route.key] || describe(route, true);
      return acc;
    }, {});
  }, [state.preloadedRoutes, describe]);
  const {
    scenes,
    shouldShowFloatOverlay,
    routeKeys,
    allRoutes,
    allDescriptors
  } = useMemo(() => {
    const allRoutes = state.routes.concat(state.preloadedRoutes);
    const scenes = [];
    const routeKeys = [];
    const allDescriptors = {
      ...preloadedDescriptors,
      ...descriptors
    };
    let shouldShowFloatOverlay = false;
    for (const route of allRoutes) {
      const descriptor = allDescriptors[route.key];
      const isPreloaded = preloadedDescriptors[route.key] !== undefined && descriptors[route.key] === undefined;
      scenes.push({
        route,
        descriptor,
        isPreloaded
      });
      routeKeys.push(route.key);
      if (!shouldShowFloatOverlay && descriptor) {
        const options = descriptor.options;
        if (options?.enableTransitions === true && options?.overlayMode === "float" && options?.overlayShown === true) {
          shouldShowFloatOverlay = true;
        }
      }
    }
    return {
      scenes,
      shouldShowFloatOverlay,
      routeKeys,
      allRoutes,
      allDescriptors
    };
  }, [state.routes, state.preloadedRoutes, descriptors, preloadedDescriptors]);

  // Get animation store maps for all routes
  const animationMaps = useMemo(() => allRoutes.map(route => AnimationStore.getAll(route.key)), [allRoutes]);
  const stackProgress = useDerivedValue(() => {
    "worklet";

    let total = 0;
    for (let i = 0; i < animationMaps.length; i++) {
      total += animationMaps[i].progress.value;
    }
    return total;
  });
  const optimisticFocusedIndex = useDerivedValue(() => {
    "worklet";

    const currentIndex = animationMaps.length - 1;
    let isAnyClosing = false;
    for (let i = 0; i < animationMaps.length; i++) {
      if (animationMaps[i].closing.value > 0) {
        isAnyClosing = true;
        break;
      }
    }
    return currentIndex - (isAnyClosing ? 1 : 0);
  });
  const focusedIndex = state.index;
  const stackContextValue = useMemo(() => ({
    flags,
    routeKeys,
    routes: allRoutes,
    descriptors: allDescriptors,
    scenes,
    focusedIndex,
    stackProgress,
    optimisticFocusedIndex
  }), [routeKeys, allRoutes, allDescriptors, scenes, focusedIndex, stackProgress, optimisticFocusedIndex, flags]);

  // DirectStack context value
  const lifecycleValue = useMemo(() => ({
    state,
    navigation,
    descriptors,
    preloadedDescriptors,
    scenes,
    focusedIndex,
    shouldShowFloatOverlay,
    stackProgress,
    optimisticFocusedIndex
  }), [state, navigation, descriptors, preloadedDescriptors, scenes, focusedIndex, shouldShowFloatOverlay, stackProgress, optimisticFocusedIndex]);
  return {
    ...lifecycleValue,
    stackContextValue
  };
}

/**
 * HOC that wraps component with DirectStack provider AND StackContext.
 * Used by native-stack which uses navigation state directly (no local route management).
 */
function withDirectStack(Component) {
  return function DirectStackProvider(props) {
    const navigatorKey = props.state.key;

    // Clean up history when navigator unmounts
    useEffect(() => {
      return () => {
        HistoryStore.clearNavigator(navigatorKey);
      };
    }, [navigatorKey]);
    const {
      stackContextValue,
      ...lifecycleValue
    } = useDirectStackValue(props);
    return /*#__PURE__*/_jsx(StackContext.Provider, {
      value: stackContextValue,
      children: /*#__PURE__*/_jsx(DirectStackContext.Provider, {
        value: lifecycleValue,
        children: /*#__PURE__*/_jsx(Component, {
          ...lifecycleValue
        })
      })
    });
  };
}
export { useDirectStackContext, withDirectStack };
