// react-native-screen-transitions/lib/module/shared/providers/stack/managed.provider.js
"use strict";

import * as React from "react";
import { useEffect, useMemo } from "react";
import { useDerivedValue } from "react-native-reanimated";
import { StackContext } from "../../hooks/navigation/use-stack";
import { AnimationStore } from "../../stores/animation.store";
import { HistoryStore } from "../../stores/history.store";
import { useStackCoreContext } from "./core.provider";
import { calculateActiveScreensLimit } from "./helpers/active-screens-limit";
import { useLocalRoutes } from "./helpers/use-local-routes";

/**
 * Props for managed stack - generic over descriptor and navigation types.
 * Defaults to base types for backward compatibility.
 */

/**
 * Context value for managed stack - generic over descriptor type.
 */
import { jsx as _jsx } from "react/jsx-runtime";
const ManagedStackContext = /*#__PURE__*/React.createContext(null);
ManagedStackContext.displayName = "ManagedStack";
function useManagedStackContext() {
  const context = React.useContext(ManagedStackContext);
  if (!context) {
    throw new Error("useManagedStackContext must be used within ManagedStackProvider");
  }
  return context;
}
function useManagedStackValue(props) {
  const {
    flags
  } = useStackCoreContext();
  const {
    state,
    handleCloseRoute,
    closingRouteKeys
  } = useLocalRoutes(props);
  const {
    scenes,
    activeScreensLimit,
    shouldShowFloatOverlay,
    routeKeys
  } = useMemo(() => {
    const scenes = [];
    const routeKeys = [];
    let shouldShowFloatOverlay = false;
    for (const route of state.routes) {
      const descriptor = state.descriptors[route.key];
      scenes.push({
        route,
        descriptor
      });
      routeKeys.push(route.key);
      if (!shouldShowFloatOverlay) {
        const options = descriptor?.options;
        shouldShowFloatOverlay = options?.overlayMode === "float" && options?.overlayShown === true;
      }
    }
    return {
      scenes,
      routeKeys,
      activeScreensLimit: calculateActiveScreensLimit(state.routes, state.descriptors),
      shouldShowFloatOverlay
    };
  }, [state.routes, state.descriptors]);

  // Get animation store maps for LOCAL routes (including closing routes)
  const animationMaps = useMemo(() => state.routes.map(route => AnimationStore.getAll(route.key)), [state.routes]);

  // Aggregated stack progress from LOCAL routes (includes closing routes)
  const stackProgress = useDerivedValue(() => {
    "worklet";

    let total = 0;
    for (let i = 0; i < animationMaps.length; i++) {
      total += animationMaps[i].progress.value;
    }
    return total;
  });

  // Optimistic focused index: accounts for closing screens
  const optimisticFocusedIndex = useDerivedValue(() => {
    "worklet";

    const currentIndex = animationMaps.length - 1;
    let isAnyClosing = false;
    for (let i = 0; i < animationMaps.length; i++) {
      if (animationMaps[i].closing.value > 0) {
        isAnyClosing = true;
        break;
      }
    }
    return currentIndex - Number(isAnyClosing);
  });
  const focusedIndex = props.state.index;

  // StackContext value - for overlays via useStack()
  const stackContextValue = useMemo(() => ({
    flags,
    routeKeys,
    routes: state.routes,
    descriptors: state.descriptors,
    scenes: scenes,
    focusedIndex,
    stackProgress,
    optimisticFocusedIndex
  }), [routeKeys, state.routes, state.descriptors, scenes, focusedIndex, stackProgress, optimisticFocusedIndex, flags]);

  // ManagedStack context value
  const lifecycleValue = useMemo(() => ({
    routes: state.routes,
    focusedIndex,
    descriptors: state.descriptors,
    closingRouteKeysShared: closingRouteKeys.shared,
    activeScreensLimit,
    handleCloseRoute,
    scenes,
    shouldShowFloatOverlay,
    stackProgress,
    optimisticFocusedIndex
  }), [state.routes, state.descriptors, focusedIndex, closingRouteKeys.shared, activeScreensLimit, handleCloseRoute, scenes, shouldShowFloatOverlay, stackProgress, optimisticFocusedIndex]);
  return {
    ...lifecycleValue,
    stackContextValue
  };
}

/**
 * HOC that wraps component with ManagedStack provider AND StackContext.
 * Used by blank-stack which manages local route state for closing animations.
 * Generic over descriptor type - defaults to BaseStackDescriptor.
 */
function withManagedStack(Component) {
  return function ManagedStackProvider(props) {
    const navigatorKey = props.state.key;

    // Clean up history when navigator unmounts
    useEffect(() => {
      return () => {
        HistoryStore.clearNavigator(navigatorKey);
      };
    }, [navigatorKey]);
    const {
      stackContextValue,
      ...lifecycleValue
    } = useManagedStackValue(props);
    return /*#__PURE__*/_jsx(StackContext.Provider, {
      value: stackContextValue,
      children: /*#__PURE__*/_jsx(ManagedStackContext.Provider, {
        value: lifecycleValue,
        children: /*#__PURE__*/_jsx(Component, {
          ...lifecycleValue
        })
      })
    });
  };
}
export { useManagedStackContext, withManagedStack };
