// expo-router/build/native-tabs/NativeBottomTabs/NativeTabsView.web.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeTabsView = NativeTabsView;
const react_tabs_1 = require("@radix-ui/react-tabs");
const react_1 = __importStar(require("react"));
const utils_1 = require("./utils");
const native_tabs_module_css_1 = __importDefault(require("../../../assets/native-tabs.module.css"));

const MAX_VISIBLE_TABS = 5;

function NativeTabsView(props) {
    const { builder, focusedIndex } = props;
    const { state, descriptors, navigation } = builder;
    const { routes } = state;
    const [showMoreScreen, setShowMoreScreen] = (0, react_1.useState)(false);
    const defaultTabName = (0, react_1.useMemo)(() => state.routes[focusedIndex]?.name ?? state.routes[0].name, []);
    const value = state.routes[focusedIndex]?.name ?? state.routes[0].name;
    const currentTabKey = state.routes[focusedIndex]?.key ?? state.routes[0].key;

    const visibleRoutes = routes.filter(({ key }) => (0, utils_1.shouldTabBeVisible)(descriptors[key].options));
    const hasOverflow = visibleRoutes.length > MAX_VISIBLE_TABS;
    const primaryRoutes = hasOverflow ? visibleRoutes.slice(0, MAX_VISIBLE_TABS - 1) : visibleRoutes;
    const overflowRoutes = hasOverflow ? visibleRoutes.slice(MAX_VISIBLE_TABS - 1) : [];

    // Check if an overflow tab is currently active
    const isOverflowTabActive = overflowRoutes.some(route => route.name === value);

    const items = primaryRoutes.map((route) => (
        <TabItem
            key={route.key}
            route={route}
            title={descriptors[route.key].options.title ?? route.name}
            badgeValue={descriptors[route.key].options.badgeValue}
            webIcon={descriptors[route.key].options.webIcon}
            webIconFamily={descriptors[route.key].options.webIconFamily}
            webIconName={descriptors[route.key].options.webIconName}
            webIconColor={descriptors[route.key].options.webIconColor}
            webIconSelectedColor={descriptors[route.key].options.webIconSelectedColor}
            webLabelColor={descriptors[route.key].options.webLabelColor}
            webLabelSelectedColor={descriptors[route.key].options.webLabelSelectedColor}
            onClick={() => setShowMoreScreen(false)}
            forceInactive={showMoreScreen}
            isActive={route.name === value && !showMoreScreen}
        />
    ));

    const children = routes
        .filter(({ key }) => (0, utils_1.shouldTabBeVisible)(descriptors[key].options))
        .map((route) => {
        return (<react_tabs_1.TabsContent key={route.name} value={route.name} className={native_tabs_module_css_1.default.tabContent} forceMount>
          {descriptors[route.key].render()}
        </react_tabs_1.TabsContent>);
    });

    const handleNavigate = (routeName) => {
        navigation.dispatch({
            type: 'JUMP_TO',
            target: state.key,
            payload: {
                name: routeName,
            },
        });
        setShowMoreScreen(false);
    };

    return (<react_tabs_1.Tabs className={native_tabs_module_css_1.default.nativeTabsContainer} defaultValue={defaultTabName} value={value} onValueChange={(newValue) => {
            handleNavigate(newValue);
        }} style={convertNativeTabsPropsToStyleVars(props, descriptors[currentTabKey]?.options)}>

      {/* More Screen - shown when More tab is active */}
      {showMoreScreen && (
          <div className={native_tabs_module_css_1.default.moreScreen}>
              <div className={native_tabs_module_css_1.default.moreScreenHeader}>
                  <h1 className={native_tabs_module_css_1.default.moreScreenTitle}>More</h1>
              </div>
              <div className={native_tabs_module_css_1.default.moreScreenContent}>
                  <div className={native_tabs_module_css_1.default.moreScreenGroup}>
                      {overflowRoutes.map((route) => (
                          <button
                              key={route.key}
                              type="button"
                              className={native_tabs_module_css_1.default.moreScreenItem}
                              onClick={() => handleNavigate(route.name)}
                          >
                              <div className={native_tabs_module_css_1.default.moreScreenItemIcon}>
                                  <OverflowTabIcon
                                      webIcon={descriptors[route.key].options.webIcon}
                                      webIconFamily={descriptors[route.key].options.webIconFamily}
                                      webIconName={descriptors[route.key].options.webIconName}
                                  />
                              </div>
                              <span className={native_tabs_module_css_1.default.moreScreenItemLabel}>
                                  {descriptors[route.key].options.title ?? route.name}
                              </span>
                              <svg className={native_tabs_module_css_1.default.moreScreenItemChevron} width="7" height="12" viewBox="0 0 7 12" fill="none">
                                  <path d="M1 1L6 6L1 11" stroke="currentColor" strokeWidth="2" strokeLinecap="round" strokeLinejoin="round"/>
                              </svg>
                          </button>
                      ))}
                  </div>
              </div>
          </div>
      )}

      {/* Tab Content - hidden when More screen is shown */}
      {!showMoreScreen && children}

      <react_tabs_1.TabsList aria-label="Main" className={native_tabs_module_css_1.default.navigationMenuRoot}>
        {items}
        {hasOverflow && (
            <button
                type="button"
                className={native_tabs_module_css_1.default.navigationMenuTrigger}
                data-state={showMoreScreen ? "active" : "inactive"}
                onClick={() => setShowMoreScreen(true)}
            >
                <span className={native_tabs_module_css_1.default.tabIcon}>
                    <svg width="18" height="18" viewBox="0 0 24 24" fill="none" stroke="currentColor" strokeWidth="2" strokeLinecap="round" strokeLinejoin="round">
                        <circle cx="12" cy="12" r="1"></circle>
                        <circle cx="19" cy="12" r="1"></circle>
                        <circle cx="5" cy="12" r="1"></circle>
                    </svg>
                </span>
                <span className={native_tabs_module_css_1.default.tabText}>More</span>
            </button>
        )}
      </react_tabs_1.TabsList>
    </react_tabs_1.Tabs>);
}

function OverflowTabIcon(props) {
    const { webIcon, webIconFamily, webIconName } = props;
    if (webIconFamily && webIconName) {
        const IconComponent = webIconFamily;
        return <IconComponent name={webIconName} size={20} />;
    } else if (webIcon) {
        return webIcon;
    }
    return null;
}

function TabItem(props) {
    const { title, badgeValue, route, webIcon, webIconFamily, webIconName, webIconColor, webIconSelectedColor, webLabelColor, webLabelSelectedColor, onClick, forceInactive, isActive } = props;
    const isBadgeEmpty = badgeValue === ' ';
    const dataState = forceInactive ? "inactive" : (isActive ? "active" : "inactive");

    // Resolve colors based on active state
    const resolvedIconColor = isActive && webIconSelectedColor ? webIconSelectedColor : webIconColor;
    const resolvedLabelColor = isActive && webLabelSelectedColor ? webLabelSelectedColor : webLabelColor;

    const iconStyle = resolvedIconColor ? { color: resolvedIconColor } : {};
    const labelStyle = resolvedLabelColor ? { color: resolvedLabelColor } : {};

    let iconElement = null;
    if (webIconFamily && webIconName) {
        const IconComponent = webIconFamily;
        iconElement = (<span className={native_tabs_module_css_1.default.tabIcon} style={iconStyle}><IconComponent name={webIconName} size={18} color={resolvedIconColor} /></span>);
    } else if (webIcon) {
        iconElement = (<span className={native_tabs_module_css_1.default.tabIcon} style={iconStyle}>{webIcon}</span>);
    }
    return (<react_tabs_1.TabsTrigger value={route.name} className={native_tabs_module_css_1.default.navigationMenuTrigger} onClick={onClick} data-state={dataState}>
      {iconElement}
      <span className={native_tabs_module_css_1.default.tabText} style={labelStyle}>{title}</span>
      {badgeValue && (<div className={`${native_tabs_module_css_1.default.tabBadge} ${isBadgeEmpty ? native_tabs_module_css_1.default.emptyTabBadge : ''}`}>
          {badgeValue}
        </div>)}
    </react_tabs_1.TabsTrigger>);
}
function convertNativeTabsPropsToStyleVars(props, currentTabOptions) {
    const vars = {};
    if (!props) {
        return vars;
    }
    const { labelStyle } = props;
    const optionsLabelStyle = currentTabOptions?.labelStyle;
    if (optionsLabelStyle?.fontFamily) {
        vars['--expo-router-tabs-font-family'] = String(optionsLabelStyle.fontFamily);
    }
    else if (labelStyle?.fontFamily) {
        vars['--expo-router-tabs-font-family'] = String(labelStyle.fontFamily);
    }
    if (optionsLabelStyle?.fontSize) {
        vars['--expo-router-tabs-font-size'] = String(optionsLabelStyle.fontSize);
    }
    else if (labelStyle?.fontSize) {
        vars['--expo-router-tabs-font-size'] = String(labelStyle.fontSize);
    }
    if (optionsLabelStyle?.fontWeight) {
        vars['--expo-router-tabs-font-weight'] = String(optionsLabelStyle.fontWeight);
    }
    else if (labelStyle?.fontWeight) {
        vars['--expo-router-tabs-font-weight'] = String(labelStyle.fontWeight);
    }
    if (optionsLabelStyle?.fontStyle) {
        vars['--expo-router-tabs-font-style'] = String(optionsLabelStyle.fontStyle);
    }
    else if (labelStyle?.fontStyle) {
        vars['--expo-router-tabs-font-style'] = String(labelStyle.fontStyle);
    }
    if (optionsLabelStyle?.color) {
        vars['--expo-router-tabs-text-color'] = String(optionsLabelStyle.color);
    }
    else if (labelStyle?.color) {
        vars['--expo-router-tabs-text-color'] = String(labelStyle.color);
    }
    if (currentTabOptions?.selectedLabelStyle?.color) {
        vars['--expo-router-tabs-active-text-color'] = String(currentTabOptions.selectedLabelStyle.color);
    }
    else if (props.tintColor) {
        vars['--expo-router-tabs-active-text-color'] = String(props.tintColor);
    }
    if (currentTabOptions?.selectedLabelStyle?.fontSize) {
        vars['--expo-router-tabs-active-font-size'] = String(currentTabOptions.selectedLabelStyle.fontSize);
    }
    if (currentTabOptions?.indicatorColor) {
        vars['--expo-router-tabs-active-background-color'] = String(currentTabOptions.indicatorColor);
    }
    else if (props.indicatorColor) {
        vars['--expo-router-tabs-active-background-color'] = String(props.indicatorColor);
    }
    if (currentTabOptions?.backgroundColor) {
        vars['--expo-router-tabs-background-color'] = String(currentTabOptions.backgroundColor);
    }
    else if (props.backgroundColor) {
        vars['--expo-router-tabs-background-color'] = String(props.backgroundColor);
    }
    if (currentTabOptions?.badgeBackgroundColor) {
        vars['--expo-router-tabs-badge-background-color'] = String(currentTabOptions.badgeBackgroundColor);
    }
    else if (props.badgeBackgroundColor) {
        vars['--expo-router-tabs-badge-background-color'] = String(props.badgeBackgroundColor);
    }
    if (currentTabOptions?.badgeTextColor) {
        vars['--expo-router-tabs-badge-text-color'] = String(currentTabOptions.badgeTextColor);
    }
    else if (props.badgeTextColor) {
        vars['--expo-router-tabs-badge-text-color'] = String(props.badgeTextColor);
    }
    return vars;
}
