// react-native-screen-transitions/lib/commonjs/shared/providers/gestures.provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useGestureContext = exports.ScreenGestureProvider = void 0;
var _reactNative = require("react-native");
var _reactNativeGestureHandler = require("react-native-gesture-handler");
var _reactNativeReanimated = require("react-native-reanimated");
var _useBuildGestures = require("../hooks/gestures/use-build-gestures");
var _useStackPointerEvents = require("../hooks/use-stack-pointer-events");
var _createProvider = _interopRequireDefault(require("../utils/create-provider"));
var _keys = require("./screen/keys.provider");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const {
  ScreenGestureProvider,
  useScreenGestureContext: useGestureContext
} = (0, _createProvider.default)("ScreenGesture", {
  guarded: false
})(({
  children
}) => {
  const {
    current
  } = (0, _keys.useKeys)();
  const pointerEvents = (0, _useStackPointerEvents.useStackPointerEvents)();
  const ancestorContext = useGestureContext();
  const scrollConfig = (0, _reactNativeReanimated.useSharedValue)(null);
  const hasGestures = current.options.gestureEnabled === true;
  const {
    panGesture,
    panGestureRef,
    nativeGesture,
    gestureAnimationValues
  } = (0, _useBuildGestures.useBuildGestures)({
    scrollConfig,
    ancestorContext
  });
  const value = {
    panGesture,
    panGestureRef,
    scrollConfig,
    nativeGesture,
    gestureAnimationValues,
    ancestorContext,
    gestureEnabled: hasGestures
  };
  return {
    value,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeGestureHandler.GestureDetector, {
      gesture: panGesture,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
        style: styles.container,
        pointerEvents: pointerEvents,
        children: children
      })
    })
  };
});
exports.useGestureContext = useGestureContext;
exports.ScreenGestureProvider = ScreenGestureProvider;
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  }
});
