// react-native-screen-transitions/lib/commonjs/shared/providers/layout-anchor.provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLayoutAnchorContext = exports.LayoutAnchorProvider = void 0;
var _reactNative = require("react-native");
var _reactNativeReanimated = require("react-native-reanimated");
var _createProvider = _interopRequireDefault(require("../utils/create-provider"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Provides a reference point for correcting bounds measurements.
 *
 * When a parent view has transforms applied (e.g., during screen transitions),
 * `measure()` returns visual positions that include those transforms. This provider
 * establishes an anchor point (typically the screen container at 0,0) and exposes
 * a `correctMeasurement` function that reverses translation and scale transforms
 * to yield the true layout position and dimensions.
 *
 * ## How it works
 *
 * 1. **Translation**: Subtract anchor's pageX/pageY offset
 * 2. **Scale**: Compare anchor's measured size to expected (screen) size to compute
 *    scale factor, then divide positions and dimensions by that factor
 */
const {
  LayoutAnchorProvider,
  useLayoutAnchorContext
} = (0, _createProvider.default)("LayoutAnchor", {
  guarded: false
})(({
  anchorRef,
  children
}) => {
  const {
    width: screenWidth,
    height: screenHeight
  } = (0, _reactNative.useWindowDimensions)();
  const correctMeasurement = measured => {
    "worklet";

    const anchor = (0, _reactNativeReanimated.measure)(anchorRef);
    if (!anchor) return measured;

    // Compute scale factor by comparing anchor size to expected screen size.
    // Anchor should be full-screen (absoluteFill), so any difference is from scale.
    const scaleX = anchor.width > 0 ? anchor.width / screenWidth : 1;
    const scaleY = anchor.height > 0 ? anchor.height / screenHeight : 1;

    // Get element position relative to anchor (removes translation)
    const relativeX = measured.pageX - anchor.pageX;
    const relativeY = measured.pageY - anchor.pageY;

    // Reverse scale: divide relative position and dimensions by scale factor
    return {
      x: measured.x,
      y: measured.y,
      width: scaleX !== 1 ? measured.width / scaleX : measured.width,
      height: scaleY !== 1 ? measured.height / scaleY : measured.height,
      pageX: scaleX !== 1 ? relativeX / scaleX : relativeX,
      pageY: scaleY !== 1 ? relativeY / scaleY : relativeY
    };
  };
  return {
    value: {
      correctMeasurement
    },
    children
  };
});
exports.useLayoutAnchorContext = useLayoutAnchorContext;
exports.LayoutAnchorProvider = LayoutAnchorProvider;
