// react-native-screen-transitions/lib/commonjs/shared/providers/register-bounds.provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RegisterBoundsProvider = void 0;
var _native = require("@react-navigation/native");
var _react = require("react");
var _reactNativeReanimated = require("react-native-reanimated");
var _useStableCallback = _interopRequireDefault(require("../hooks/use-stable-callback"));
var _useStableCallbackValue = _interopRequireDefault(require("../hooks/use-stable-callback-value"));
var _animation = require("../stores/animation.store");
var _bounds = require("../stores/bounds.store");
var _styles = require("../utils/bounds/helpers/styles");
var _createProvider = _interopRequireDefault(require("../utils/create-provider"));
var _layoutAnchor = require("./layout-anchor.provider");
var _keys = require("./screen/keys.provider");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Builds the full ancestor key chain for nested navigators.
 * Returns an array of screen keys from immediate parent to root.
 * [parentKey, grandparentKey, greatGrandparentKey, ...]
 */
const getAncestorKeys = current => {
  const ancestors = [];
  const nav = current.navigation;

  // Safety check for navigators without getParent
  if (typeof nav?.getParent !== "function") {
    return ancestors;
  }
  let parent = nav.getParent();
  while (parent) {
    const state = parent.getState();
    if (state?.routes && state.index !== undefined) {
      const focusedRoute = state.routes[state.index];
      if (focusedRoute?.key) {
        ancestors.push(focusedRoute.key);
      }
    }
    parent = parent.getParent();
  }
  return ancestors;
};

/**
 * Handles initial layout measurement for destination elements.
 * Only measures if an animation is in progress (local or parent).
 */
const useInitialLayoutHandler = params => {
  const {
    sharedBoundTag,
    currentScreenKey,
    ancestorKeys,
    maybeMeasureAndStore
  } = params;
  const isAnimating = _animation.AnimationStore.getAnimation(currentScreenKey, "animating");

  // Check if any ancestor is animating
  const ancestorAnimations = ancestorKeys.map(key => _animation.AnimationStore.getAnimation(key, "animating"));
  const hasMeasuredOnLayout = (0, _reactNativeReanimated.useSharedValue)(false);
  return (0, _react.useCallback)(() => {
    "worklet";

    if (!sharedBoundTag || hasMeasuredOnLayout.get()) return;

    // Check if current or any ancestor is animating
    let isAnyAnimating = isAnimating.get();
    for (let i = 0; i < ancestorAnimations.length; i++) {
      if (ancestorAnimations[i].get()) {
        isAnyAnimating = 1;
        break;
      }
    }
    if (!isAnyAnimating) return;
    maybeMeasureAndStore({
      shouldSetSource: false,
      shouldSetDestination: true
    });
    hasMeasuredOnLayout.set(true);
  }, [sharedBoundTag, hasMeasuredOnLayout, isAnimating, ancestorAnimations, maybeMeasureAndStore]);
};

/**
 * Measures non-pressable elements when screen becomes blurred.
 * Captures bounds right before transition starts.
 */
/**
 * Measures non-pressable elements when screen becomes blurred.
 * Captures bounds right before transition starts.
 */
const useBlurMeasurement = params => {
  const {
    current
  } = (0, _keys.useKeys)();
  const {
    sharedBoundTag,
    ancestorKeys,
    maybeMeasureAndStore
  } = params;
  const hasCapturedSource = (0, _react.useRef)(false);
  const ancestorClosing = [current.route.key, ...ancestorKeys].map(key => _animation.AnimationStore.getAnimation(key, "closing"));
  const maybeMeasureOnBlur = (0, _useStableCallbackValue.default)(() => {
    "worklet";

    //.some doesnt work here apparently... :-(
    for (const closing of ancestorClosing) {
      if (closing.get()) return;
    }
    maybeMeasureAndStore({
      shouldSetSource: true
    });
  });
  (0, _native.useFocusEffect)((0, _react.useCallback)(() => {
    hasCapturedSource.current = false;
    return () => {
      if (!sharedBoundTag || hasCapturedSource.current) return;
      (0, _reactNativeReanimated.runOnUI)(maybeMeasureOnBlur)();
    };
  }, [sharedBoundTag, maybeMeasureOnBlur]));
  return {
    markSourceCaptured: () => {
      hasCapturedSource.current = true;
    }
  };
};

/**
 * Syncs child measurements when parent updates.
 */
const useParentSyncReaction = params => {
  const {
    parentContext,
    maybeMeasureAndStore
  } = params;
  (0, _reactNativeReanimated.useAnimatedReaction)(() => parentContext?.updateSignal.get(), value => {
    "worklet";

    if (value === 0 || value === undefined) return;
    maybeMeasureAndStore();
  });
};
const {
  RegisterBoundsProvider,
  useRegisterBoundsContext
} = (0, _createProvider.default)("RegisterBounds", {
  guarded: false
})(({
  style,
  onPress,
  sharedBoundTag,
  animatedRef,
  children
}) => {
  const {
    current
  } = (0, _keys.useKeys)();
  const currentScreenKey = current.route.key;
  const ancestorKeys = (0, _react.useMemo)(() => getAncestorKeys(current), [current]);
  const layoutAnchor = (0, _layoutAnchor.useLayoutAnchorContext)();

  // Context & signals
  const parentContext = useRegisterBoundsContext();
  const ownSignal = (0, _reactNativeReanimated.useSharedValue)(0);
  const updateSignal = parentContext?.updateSignal ?? ownSignal;
  const isAnimating = _animation.AnimationStore.getAnimation(currentScreenKey, "animating");
  const preparedStyles = (0, _react.useMemo)(() => (0, _styles.prepareStyleForBounds)(style), [style]);
  const emitUpdate = (0, _useStableCallbackValue.default)(() => {
    "worklet";

    const isRoot = !parentContext;
    if (isRoot) updateSignal.set(updateSignal.get() + 1);
  });
  const maybeMeasureAndStore = (0, _useStableCallbackValue.default)(({
    onPress,
    shouldSetSource,
    shouldSetDestination
  } = {}) => {
    "worklet";

    if (!sharedBoundTag) return;
    const measured = (0, _reactNativeReanimated.measure)(animatedRef);
    if (!measured) return;

    // Correct for parent transforms (e.g., when parent screen is animating)
    const correctedMeasured = layoutAnchor ? layoutAnchor.correctMeasurement(measured) : measured;
    emitUpdate();
    _bounds.BoundStore.registerSnapshot(sharedBoundTag, currentScreenKey, correctedMeasured, preparedStyles, ancestorKeys);
    if (shouldSetSource) {
      if (isAnimating.get()) {
        // If animation is already in progress,
        // lets use the existing measuremenets.
        const existing = _bounds.BoundStore.getSnapshot(sharedBoundTag, currentScreenKey);
        if (existing) {
          _bounds.BoundStore.setLinkSource(sharedBoundTag, currentScreenKey, existing.bounds, preparedStyles, ancestorKeys);
        }
        return;
      }
      _bounds.BoundStore.setLinkSource(sharedBoundTag, currentScreenKey, correctedMeasured, preparedStyles, ancestorKeys);
    }

    // Set as destination (on mount during animation)
    if (shouldSetDestination) {
      _bounds.BoundStore.setLinkDestination(sharedBoundTag, currentScreenKey, correctedMeasured, preparedStyles, ancestorKeys);
    }
    if (onPress) (0, _reactNativeReanimated.runOnJS)(onPress)();
  });
  const handleInitialLayout = useInitialLayoutHandler({
    sharedBoundTag,
    currentScreenKey,
    ancestorKeys,
    maybeMeasureAndStore
  });

  // Side effects
  const {
    markSourceCaptured
  } = useBlurMeasurement({
    sharedBoundTag,
    maybeMeasureAndStore,
    ancestorKeys
  });
  useParentSyncReaction({
    parentContext,
    maybeMeasureAndStore
  });
  const captureActiveOnPress = (0, _useStableCallback.default)(() => {
    if (!sharedBoundTag) {
      onPress?.();
      return;
    }
    (0, _reactNativeReanimated.runOnUI)(maybeMeasureAndStore)({
      onPress,
      shouldSetSource: true
    });
    markSourceCaptured();
  });
  return {
    value: {
      updateSignal
    },
    children: children({
      handleInitialLayout,
      captureActiveOnPress
    })
  };
});
exports.RegisterBoundsProvider = RegisterBoundsProvider;
