// react-native-screen-transitions/lib/module/shared/components/integrations/masked-view.js
"use strict";

import { StyleSheet, View } from "react-native";
import { CONTAINER_STYLE_ID, MASK_STYLE_ID } from "../../constants";
import { createTransitionAwareComponent } from "../create-transition-aware-component";
import { jsx as _jsx } from "react/jsx-runtime";
const TransitionView = createTransitionAwareComponent(View);
let LazyMaskedView = View;
try {
  LazyMaskedView = require("@react-native-masked-view/masked-view").default;
} catch (_) {
  // noop
}
export default function MaskedView({
  children,
  style: userStyles = {}
}) {
  if (LazyMaskedView === View) {
    return children;
  }
  return /*#__PURE__*/_jsx(LazyMaskedView, {
    style: styles.root
    // @ts-expect-error
    ,
    maskElement: /*#__PURE__*/_jsx(TransitionView, {
      styleId: MASK_STYLE_ID,
      style: styles.rootMask
    }),
    children: /*#__PURE__*/_jsx(TransitionView, {
      styleId: CONTAINER_STYLE_ID,
      style: [styles.rootContainer, userStyles],
      children: children
    })
  });
}
const styles = StyleSheet.create({
  root: {
    flex: 1
  },
  rootMask: {
    backgroundColor: "white"
  },
  rootContainer: {
    flex: 1
  }
});
