// react-native-screen-transitions/lib/module/shared/components/overlay/variations/overlay-host.js
"use strict";

import { NavigationContext, NavigationRouteContext } from "@react-navigation/native";
import { memo, useMemo } from "react";
import { Animated, StyleSheet, useWindowDimensions, View } from "react-native";
import { useDerivedValue } from "react-native-reanimated";
import { useSafeAreaInsets } from "react-native-safe-area-context";
import { useScreenAnimation } from "../../../hooks/animation/use-screen-animation";
import { useScreenState } from "../../../hooks/navigation/use-screen-state";
import { jsx as _jsx } from "react/jsx-runtime";
export const OverlayHost = /*#__PURE__*/memo(function OverlayHost({
  scene
}) {
  const OverlayComponent = scene.descriptor.options.overlay;
  const screen = useWindowDimensions();
  const insets = useSafeAreaInsets();
  const screenAnimation = useScreenAnimation();
  const screenState = useScreenState();
  const relativeProgress = useDerivedValue(() => {
    "worklet";

    return screenAnimation.value.stackProgress;
  });
  const overlayAnimation = useDerivedValue(() => ({
    progress: relativeProgress.value,
    layouts: {
      screen
    },
    insets
  }));
  const overlayProps = useMemo(() => ({
    ...screenState,
    progress: relativeProgress,
    overlayAnimation,
    //Deprecated
    screenAnimation //Deprecated
  }), [relativeProgress, overlayAnimation, screenAnimation, screenState]);
  if (!OverlayComponent) {
    return null;
  }
  return /*#__PURE__*/_jsx(Animated.View, {
    pointerEvents: "box-none",
    style: [styles.container, styles.floating, styles.absolute],
    children: /*#__PURE__*/_jsx(NavigationContext.Provider, {
      value: scene.descriptor.navigation,
      children: /*#__PURE__*/_jsx(NavigationRouteContext.Provider, {
        value: scene.route,
        children: /*#__PURE__*/_jsx(View, {
          pointerEvents: "box-none",
          style: styles.overlay,
          children: /*#__PURE__*/_jsx(OverlayComponent, {
            ...overlayProps
          })
        })
      })
    })
  });
});
const styles = StyleSheet.create({
  overlay: {
    ...StyleSheet.absoluteFillObject,
    zIndex: 1
  },
  container: {
    flex: 1
  },
  absolute: StyleSheet.absoluteFillObject,
  floating: {
    zIndex: 1000
  }
});
