// react-native-screen-transitions/lib/commonjs/component-stack/components/component-screen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentScreen = void 0;
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _layoutAnchor = require("../../shared/providers/layout-anchor.provider");
var _animation = require("../../shared/stores/animation.store");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const POINT_NONE = "none";
const POINT_BOX_NONE = "box-none";
const ComponentScreen = ({
  routeKey,
  children
}) => {
  const sceneClosing = _animation.AnimationStore.getAnimation(routeKey, "closing");
  const screenRef = (0, _reactNativeReanimated.useAnimatedRef)();
  const animatedProps = (0, _reactNativeReanimated.useAnimatedProps)(() => {
    return {
      pointerEvents: sceneClosing.get() ? POINT_NONE : POINT_BOX_NONE
    };
  });
  const ComponentScreenComponent = _reactNativeReanimated.default.View;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ComponentScreenComponent, {
    ref: screenRef,
    style: _reactNative.StyleSheet.absoluteFill,
    animatedProps: animatedProps,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_layoutAnchor.LayoutAnchorProvider, {
      anchorRef: screenRef,
      children: children
    })
  });
};
exports.ComponentScreen = ComponentScreen;
