// react-native-screen-transitions/lib/module/shared/providers/gestures.provider.js
"use strict";

import { StyleSheet, View } from "react-native";
import { GestureDetector } from "react-native-gesture-handler";
import { useSharedValue } from "react-native-reanimated";
import { useBuildGestures } from "../hooks/gestures/use-build-gestures";
import { useStackPointerEvents } from "../hooks/use-stack-pointer-events";
import createProvider from "../utils/create-provider";
import { useKeys } from "./screen/keys.provider";
import { jsx as _jsx } from "react/jsx-runtime";
export const {
  ScreenGestureProvider,
  useScreenGestureContext: useGestureContext
} = createProvider("ScreenGesture", {
  guarded: false
})(({
  children
}) => {
  const {
    current
  } = useKeys();
  const pointerEvents = useStackPointerEvents();
  const ancestorContext = useGestureContext();
  const scrollConfig = useSharedValue(null);
  const hasGestures = current.options.gestureEnabled === true;
  const {
    panGesture,
    panGestureRef,
    nativeGesture,
    gestureAnimationValues
  } = useBuildGestures({
    scrollConfig,
    ancestorContext
  });
  const value = {
    panGesture,
    panGestureRef,
    scrollConfig,
    nativeGesture,
    gestureAnimationValues,
    ancestorContext,
    gestureEnabled: hasGestures
  };
  return {
    value,
    children: /*#__PURE__*/_jsx(GestureDetector, {
      gesture: panGesture,
      children: /*#__PURE__*/_jsx(View, {
        style: styles.container,
        pointerEvents: pointerEvents,
        children: children
      })
    })
  };
});
const styles = StyleSheet.create({
  container: {
    flex: 1
  }
});
