// react-native-screen-transitions/lib/commonjs/shared/components/overlay/variations/overlay-host.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OverlayHost = void 0;
var _native = require("@react-navigation/native");
var _react = require("react");
var _reactNative = require("react-native");
var _reactNativeReanimated = require("react-native-reanimated");
var _reactNativeSafeAreaContext = require("react-native-safe-area-context");
var _useScreenAnimation = require("../../../hooks/animation/use-screen-animation");
var _useScreenState = require("../../../hooks/navigation/use-screen-state");
var _jsxRuntime = require("react/jsx-runtime");
const OverlayHost = exports.OverlayHost = /*#__PURE__*/(0, _react.memo)(function OverlayHost({
  scene
}) {
  const OverlayComponent = scene.descriptor.options.overlay;
  const screen = (0, _reactNative.useWindowDimensions)();
  const insets = (0, _reactNativeSafeAreaContext.useSafeAreaInsets)();
  const screenAnimation = (0, _useScreenAnimation.useScreenAnimation)();
  const screenState = (0, _useScreenState.useScreenState)();
  const relativeProgress = (0, _reactNativeReanimated.useDerivedValue)(() => {
    "worklet";

    return screenAnimation.value.stackProgress;
  });
  const overlayAnimation = (0, _reactNativeReanimated.useDerivedValue)(() => ({
    progress: relativeProgress.value,
    layouts: {
      screen
    },
    insets
  }));
  const overlayProps = (0, _react.useMemo)(() => ({
    ...screenState,
    progress: relativeProgress,
    overlayAnimation,
    //Deprecated
    screenAnimation //Deprecated
  }), [relativeProgress, overlayAnimation, screenAnimation, screenState]);
  if (!OverlayComponent) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Animated.View, {
    pointerEvents: "box-none",
    style: [styles.container, styles.floating, styles.absolute],
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_native.NavigationContext.Provider, {
      value: scene.descriptor.navigation,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_native.NavigationRouteContext.Provider, {
        value: scene.route,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
          pointerEvents: "box-none",
          style: styles.overlay,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(OverlayComponent, {
            ...overlayProps
          })
        })
      })
    })
  });
});
const styles = _reactNative.StyleSheet.create({
  overlay: {
    ..._reactNative.StyleSheet.absoluteFillObject,
    zIndex: 1
  },
  container: {
    flex: 1
  },
  absolute: _reactNative.StyleSheet.absoluteFillObject,
  floating: {
    zIndex: 1000
  }
});
