// react-native-screen-transitions/lib/commonjs/shared/utils/bounds/helpers/style-composers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.composeContentStyle = composeContentStyle;
exports.composeSizeAbsolute = composeSizeAbsolute;
exports.composeSizeRelative = composeSizeRelative;
exports.composeTransformAbsolute = composeTransformAbsolute;
exports.composeTransformRelative = composeTransformRelative;
var _reactNativeReanimated = require("react-native-reanimated");
/**
 * Element-level (relative) params shared by size/transform composers.
 * - start/end: absolute window bounds of the element in previous/next phases
 * - geometry: relative deltas and scales between start/end (dx, dy, scaleX, scaleY, ...)
 * - interp: function to interpolate between numbers using the correct progress range
 */

/**
 * Screen-level content transform params (for aligning destination bound to source).
 * - start/end: absolute window bounds for the shared id (source/destination)
 * - geometry: precomputed screen-level tx/ty/sx/sy plus ranges/entering
 * - interp: function to interpolate between numbers using the correct progress range
 */

function composeSizeAbsolute(params) {
  "worklet";

  const {
    start,
    end,
    geometry,
    progress,
    ranges,
    computeOptions
  } = params;
  const width = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.width, end.width]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.width, start.width]);
  const height = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.height, end.height]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.height, start.height]);
  const translateX = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.pageX, end.pageX]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.pageX, start.pageX]);
  const translateY = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.pageY, end.pageY]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.pageY, start.pageY]);
  if (computeOptions.raw) {
    return {
      width,
      height,
      translateX,
      translateY
    };
  }
  return {
    width,
    height,
    transform: [{
      translateX
    }, {
      translateY
    }]
  };
}
function composeSizeRelative(params) {
  "worklet";

  const {
    start,
    end,
    geometry,
    progress,
    ranges,
    computeOptions
  } = params;
  const translateX = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.dx, 0]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [0, -geometry.dx]);
  const translateY = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.dy, 0]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [0, -geometry.dy]);
  const width = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.width, end.width]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.width, start.width]);
  const height = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.height, end.height]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.height, start.height]);
  if (computeOptions.raw) {
    return {
      translateX,
      translateY,
      width,
      height
    };
  }
  return {
    transform: [{
      translateX
    }, {
      translateY
    }],
    width,
    height
  };
}
function composeTransformAbsolute(params) {
  "worklet";

  const {
    start,
    end,
    geometry,
    progress,
    ranges,
    computeOptions
  } = params;
  const translateX = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.pageX, end.pageX]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.pageX, start.pageX]);
  const translateY = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [start.pageY, end.pageY]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [end.pageY, start.pageY]);
  const scaleX = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.scaleX, 1]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [1, 1 / geometry.scaleX]);
  const scaleY = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.scaleY, 1]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [1, 1 / geometry.scaleY]);
  if (computeOptions.raw) {
    return {
      translateX,
      translateY,
      scaleX,
      scaleY
    };
  }
  return {
    transform: [{
      translateX
    }, {
      translateY
    }, {
      scaleX
    }, {
      scaleY
    }]
  };
}
function composeTransformRelative(params) {
  "worklet";

  const {
    geometry,
    computeOptions,
    progress,
    ranges
  } = params;
  const translateX = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.dx, 0]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [0, -geometry.dx]);
  const translateY = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.dy, 0]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [0, -geometry.dy]);
  const scaleX = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.scaleX, 1]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [1, 1 / geometry.scaleX]);
  const scaleY = geometry.entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [geometry.scaleY, 1]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [1, 1 / geometry.scaleY]);
  if (computeOptions.raw) {
    return {
      translateX,
      translateY,
      scaleX,
      scaleY
    };
  }
  return {
    transform: [{
      translateX: computeOptions.gestures?.x ?? 0
    }, {
      translateY: computeOptions.gestures?.y ?? 0
    }, {
      translateX
    }, {
      translateY
    }, {
      scaleX
    }, {
      scaleY
    }]
  };
}
function composeContentStyle(params) {
  "worklet";

  const {
    geometry,
    progress,
    ranges,
    computeOptions: {
      raw
    }
  } = params;
  const {
    s,
    tx,
    ty,
    entering
  } = geometry;
  const translateX = entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [tx, 0]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [0, tx]);
  const translateY = entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [ty, 0]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [0, ty]);
  const scale = entering ? (0, _reactNativeReanimated.interpolate)(progress, ranges, [s, 1]) : (0, _reactNativeReanimated.interpolate)(progress, ranges, [1, s]);
  if (raw) {
    return {
      translateX,
      translateY,
      scale
    };
  }
  return {
    transform: [{
      translateX
    }, {
      translateY
    }, {
      scale
    }]
  };
}
