// react-native-screen-transitions/lib/commonjs/shared/providers/stack/helpers/use-local-routes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLocalRoutes = void 0;
var _native = require("@react-navigation/native");
var _react = require("react");
var _useClosingRouteKeys = require("../../../hooks/navigation/use-closing-route-keys");
var _usePrevious = require("../../../hooks/navigation/use-previous");
var _useStableCallback = _interopRequireDefault(require("../../../hooks/use-stable-callback"));
var _alignRoutesWithLatest = require("../../../utils/navigation/align-routes-with-latest");
var _areDescriptorsEqual = require("../../../utils/navigation/are-descriptors-equal");
var _haveSameRouteKeys = require("../../../utils/navigation/have-same-route-keys");
var _routesAreIdentical = require("../../../utils/navigation/routes-are-identical");
var _syncRoutesWithRemoved = require("../../../utils/navigation/sync-routes-with-removed");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const useLocalRoutes = props => {
  const previousRoutes = (0, _usePrevious.usePrevious)(props.state.routes) ?? [];
  const closingRouteKeys = (0, _useClosingRouteKeys.useClosingRouteKeys)();
  const [localState, setLocalState] = (0, _react.useState)(() => ({
    routes: props.state.routes,
    descriptors: props.descriptors
  }));
  (0, _react.useLayoutEffect)(() => {
    const nextRoutesSnapshot = props.state.routes;
    const previousRoutesSnapshot = previousRoutes;
    setLocalState(current => {
      const routeKeysUnchanged = (0, _haveSameRouteKeys.haveSameRouteKeys)(previousRoutesSnapshot, nextRoutesSnapshot);
      let derivedRoutes;
      let derivedDescriptors;
      if (routeKeysUnchanged) {
        const result = (0, _alignRoutesWithLatest.alignRoutesWithLatest)(current.routes, current.descriptors, nextRoutesSnapshot, props.descriptors);
        derivedRoutes = result.routes;
        derivedDescriptors = result.descriptors;
      } else {
        const fallbackRoutes = previousRoutesSnapshot.length > 0 ? previousRoutesSnapshot : current.routes;
        const result = (0, _syncRoutesWithRemoved.syncRoutesWithRemoved)({
          prevRoutes: fallbackRoutes,
          prevDescriptors: current.descriptors,
          nextRoutes: nextRoutesSnapshot,
          nextDescriptors: props.descriptors,
          closingRouteKeys
        });
        derivedRoutes = result.routes;
        derivedDescriptors = result.descriptors;
      }
      const routesChanged = !(0, _routesAreIdentical.routesAreIdentical)(current.routes, derivedRoutes);
      const descriptorsChanged = !(0, _areDescriptorsEqual.areDescriptorsEqual)(current.descriptors, derivedDescriptors);
      if (!routesChanged && !descriptorsChanged) {
        return current;
      }
      return {
        routes: routesChanged ? derivedRoutes : current.routes,
        descriptors: descriptorsChanged ? derivedDescriptors : current.descriptors
      };
    });
  }, [props.state.routes, props.descriptors, previousRoutes, closingRouteKeys]);
  const handleCloseRoute = (0, _useStableCallback.default)(({
    route
  }) => {
    if (props.state.routes.some(r => r.key === route.key)) {
      props.navigation.dispatch({
        ..._native.StackActions.pop(),
        source: route.key,
        target: props.state.key
      });
      return;
    }
    closingRouteKeys.remove(route.key);
    setLocalState(current => {
      if (!current.routes.some(candidate => candidate.key === route.key)) {
        return current;
      }
      const nextRoutes = current.routes.filter(candidate => candidate.key !== route.key);
      const nextDescriptors = {
        ...current.descriptors
      };
      delete nextDescriptors[route.key];
      return {
        routes: nextRoutes,
        descriptors: nextDescriptors
      };
    });
  });
  return {
    state: localState,
    handleCloseRoute,
    closingRouteKeys
  };
};
exports.useLocalRoutes = useLocalRoutes;
