// react-native-screen-transitions/lib/module/component-stack/components/component-screen.js
"use strict";

import { StyleSheet } from "react-native";
import Animated, { useAnimatedProps, useAnimatedRef } from "react-native-reanimated";
import { LayoutAnchorProvider } from "../../shared/providers/layout-anchor.provider";
import { AnimationStore } from "../../shared/stores/animation.store";
import { jsx as _jsx } from "react/jsx-runtime";
const POINT_NONE = "none";
const POINT_BOX_NONE = "box-none";
export const ComponentScreen = ({
  routeKey,
  children
}) => {
  const sceneClosing = AnimationStore.getAnimation(routeKey, "closing");
  const screenRef = useAnimatedRef();
  const animatedProps = useAnimatedProps(() => {
    return {
      pointerEvents: sceneClosing.get() ? POINT_NONE : POINT_BOX_NONE
    };
  });
  const ComponentScreenComponent = Animated.View;
  return /*#__PURE__*/_jsx(ComponentScreenComponent, {
    ref: screenRef,
    style: StyleSheet.absoluteFill,
    animatedProps: animatedProps,
    children: /*#__PURE__*/_jsx(LayoutAnchorProvider, {
      anchorRef: screenRef,
      children: children
    })
  });
};
