// react-native-screen-transitions/lib/module/component-stack/components/stack-view.js
"use strict";

import { NavigationContext, NavigationRouteContext } from "@react-navigation/native";
import * as React from "react";
import { Fragment } from "react";
import { Overlay } from "../../shared/components/overlay";
import { ScreenComposer } from "../../shared/providers/screen/screen-composer";
import { withStackCore } from "../../shared/providers/stack/core.provider";
import { withManagedStack } from "../../shared/providers/stack/managed.provider";
import { StackType } from "../../shared/types/stack.types";
import { ComponentScreen } from "./component-screen";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const SceneView = /*#__PURE__*/React.memo(function SceneView({
  descriptor
}) {
  const {
    route,
    navigation,
    render
  } = descriptor;
  return /*#__PURE__*/_jsx(NavigationContext.Provider, {
    value: navigation,
    children: /*#__PURE__*/_jsxs(NavigationRouteContext.Provider, {
      value: route,
      children: [descriptor.options.overlayMode === "screen" && /*#__PURE__*/_jsx(Overlay.Screen, {}), render()]
    })
  });
});
export const StackView = withStackCore({
  TRANSITIONS_ALWAYS_ON: true,
  STACK_TYPE: StackType.COMPONENT
}, withManagedStack(({
  scenes,
  shouldShowFloatOverlay
}) => {
  return /*#__PURE__*/_jsxs(Fragment, {
    children: [shouldShowFloatOverlay ? /*#__PURE__*/_jsx(Overlay.Float, {}) : null, scenes.map((scene, sceneIndex) => {
      const descriptor = scene.descriptor;
      const route = scene.route;
      const previousDescriptor = scenes[sceneIndex - 1]?.descriptor ?? undefined;
      const nextDescriptor = scenes[sceneIndex + 1]?.descriptor ?? undefined;
      return /*#__PURE__*/_jsx(ComponentScreen, {
        routeKey: route.key,
        children: /*#__PURE__*/_jsx(ScreenComposer, {
          previous: previousDescriptor,
          current: descriptor,
          next: nextDescriptor,
          children: /*#__PURE__*/_jsx(SceneView, {
            descriptor: descriptor
          }, route.key)
        })
      }, route.key);
    })]
  });
}));
