// react-native-screen-transitions/lib/module/component-stack/navigators/create-component-stack-navigator.js
"use strict";

import { createNavigatorFactory, NavigationContainer, NavigationIndependentTree, StackRouter, useNavigationBuilder } from "@react-navigation/native";
import * as React from "react";
import { StackView } from "../components/stack-view";
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * Context to detect if we're inside a ComponentStack.
 * Used to avoid double-wrapping nested ComponentStacks in NavigationIndependentTree.
 */
const ComponentStackContext = /*#__PURE__*/React.createContext(false);
ComponentStackContext.displayName = "ComponentStackContext";
const ComponentStackNavigatorInner = /*#__PURE__*/React.memo(function ComponentStackNavigatorInner({
  id,
  initialRouteName,
  children,
  layout,
  screenListeners,
  screenOptions,
  screenLayout,
  ...rest
}) {
  const {
    state,
    describe,
    descriptors,
    navigation,
    NavigationContent
  } = useNavigationBuilder(StackRouter, {
    id,
    initialRouteName,
    children,
    layout,
    screenListeners,
    screenOptions,
    screenLayout
  });
  return /*#__PURE__*/_jsx(NavigationContent, {
    children: /*#__PURE__*/_jsx(StackView, {
      ...rest,
      state: state,
      navigation: navigation,
      descriptors: descriptors,
      describe: describe
    })
  });
});

/**
 * ComponentStackNavigator conditionally wraps in NavigationIndependentTree.
 * - If already inside a ComponentStack (nested), just render the inner navigator
 *   so it participates in the parent's navigation tree and goBack() works.
 * - If top-level, wrap in NavigationIndependentTree + NavigationContainer
 *   to isolate from Expo Router / React Navigation.
 */
function IsolatedComponentStackNavigator(props) {
  const isNested = React.useContext(ComponentStackContext);
  if (isNested) {
    return /*#__PURE__*/_jsx(ComponentStackNavigatorInner, {
      ...props
    });
  }
  return /*#__PURE__*/_jsx(NavigationIndependentTree, {
    children: /*#__PURE__*/_jsx(NavigationContainer, {
      children: /*#__PURE__*/_jsx(ComponentStackContext.Provider, {
        value: true,
        children: /*#__PURE__*/_jsx(ComponentStackNavigatorInner, {
          ...props
        })
      })
    })
  });
}
export function createComponentStackNavigator(config) {
  return createNavigatorFactory(IsolatedComponentStackNavigator)(config);
}
