// @anythingai/app/build/chunk-KQ66NBHH.js
'use strict';var p=require('@apollo/client');function _interopNamespace(e){if(e&&e.__esModule)return e;var n=Object.create(null);if(e){Object.keys(e).forEach(function(k){if(k!=='default'){var d=Object.getOwnPropertyDescriptor(e,k);Object.defineProperty(n,k,d.get?d:{enumerable:true,get:function(){return e[k]}});}})}n.default=e;return Object.freeze(n)}var p__namespace=/*#__PURE__*/_interopNamespace(p);var i={};var c=(n=>(n.ANTHROPIC_CLAUDE="ANTHROPIC_CLAUDE",n.ANTHROPIC_CLAUDE_HAIKU="ANTHROPIC_CLAUDE_HAIKU",n.ANTHROPIC_CLAUDE_HAIKU_4_5="ANTHROPIC_CLAUDE_HAIKU_4_5",n.ANTHROPIC_CLAUDE_OPUS_4_1="ANTHROPIC_CLAUDE_OPUS_4_1",n.ANTHROPIC_CLAUDE_OPUS_4_5="ANTHROPIC_CLAUDE_OPUS_4_5",n.ANTHROPIC_CLAUDE_SONNET="ANTHROPIC_CLAUDE_SONNET",n.ANTHROPIC_CLAUDE_SONNET_3_5="ANTHROPIC_CLAUDE_SONNET_3_5",n.ANTHROPIC_CLAUDE_SONNET_4="ANTHROPIC_CLAUDE_SONNET_4",n.ANTHROPIC_CLAUDE_SONNET_4_5="ANTHROPIC_CLAUDE_SONNET_4_5",n.CHAT_GPT="CHAT_GPT",n.COHERE_COMMAND_R_PLUS="COHERE_COMMAND_R_PLUS",n.DALL_E_3="DALL_E_3",n.DEEPSEEK_V3="DEEPSEEK_V3",n.FAST_LIGHTNING_SDXL="FAST_LIGHTNING_SDXL",n.GOOGLE_GEMINI="GOOGLE_GEMINI",n.GOOGLE_GEMINI_1_5="GOOGLE_GEMINI_1_5",n.GOOGLE_GEMINI_1_5_FLASH="GOOGLE_GEMINI_1_5_FLASH",n.GOOGLE_GEMINI_2_0_FLASH="GOOGLE_GEMINI_2_0_FLASH",n.GOOGLE_GEMINI_2_5_FLASH="GOOGLE_GEMINI_2_5_FLASH",n.GOOGLE_GEMINI_2_5_FLASH_IMAGE="GOOGLE_GEMINI_2_5_FLASH_IMAGE",n.GOOGLE_GEMINI_2_5_FLASH_LITE="GOOGLE_GEMINI_2_5_FLASH_LITE",n.GOOGLE_GEMINI_2_5_PRO="GOOGLE_GEMINI_2_5_PRO",n.GOOGLE_GEMINI_3_0_FLASH="GOOGLE_GEMINI_3_0_FLASH",n.GOOGLE_GEMINI_3_0_PRO="GOOGLE_GEMINI_3_0_PRO",n.GPT_IMAGE_1="GPT_IMAGE_1",n.GPT_IMAGE_1_5="GPT_IMAGE_1_5",n.GROQ="GROQ",n.GROQ_KIMI_K2="GROQ_KIMI_K2",n.MORPH="MORPH",n.RAPID_API="RAPID_API",n.REKA="REKA",n.RELACE="RELACE",n.STABLE_DIFFUSION_V3_MEDIUM="STABLE_DIFFUSION_V3_MEDIUM",n.XAI_GROK_4="XAI_GROK_4",n))(c||{});var S=(o=>(o.ANIME_AND_MANGA="ANIME_AND_MANGA",o.ANTHROPIC_CLAUDE="ANTHROPIC_CLAUDE",o.ANTHROPIC_CLAUDE_HAIKU="ANTHROPIC_CLAUDE_HAIKU",o.ANTHROPIC_CLAUDE_OPUS_4_1="ANTHROPIC_CLAUDE_OPUS_4_1",o.ANTHROPIC_CLAUDE_SONNET="ANTHROPIC_CLAUDE_SONNET",o.ANTHROPIC_CLAUDE_SONNET_3_5="ANTHROPIC_CLAUDE_SONNET_3_5",o.ANTHROPIC_CLAUDE_SONNET_4="ANTHROPIC_CLAUDE_SONNET_4",o.ANTHROPIC_CLAUDE_SONNET_4_5="ANTHROPIC_CLAUDE_SONNET_4_5",o.BASKETBALL="BASKETBALL",o.BOOK_SEARCH="BOOK_SEARCH",o.CHAT_GPT="CHAT_GPT",o.CODE_RUNNER="CODE_RUNNER",o.COHERE_COMMAND_R_PLUS="COHERE_COMMAND_R_PLUS",o.DALL_E_3="DALL_E_3",o.DOMAIN_WHOIS="DOMAIN_WHOIS",o.EMAIL_CHECK="EMAIL_CHECK",o.EMAIL_CHECK_2="EMAIL_CHECK_2",o.FILE_CONVERTER="FILE_CONVERTER",o.GOOGLE_GEMINI="GOOGLE_GEMINI",o.GOOGLE_GEMINI_1_5="GOOGLE_GEMINI_1_5",o.GOOGLE_GEMINI_1_5_FLASH="GOOGLE_GEMINI_1_5_FLASH",o.GOOGLE_GEMINI_2_5_FLASH="GOOGLE_GEMINI_2_5_FLASH",o.GOOGLE_GEMINI_2_5_PRO="GOOGLE_GEMINI_2_5_PRO",o.GOOGLE_GEMINI_3_0_PRO="GOOGLE_GEMINI_3_0_PRO",o.GOOGLE_PLACE_AUTOCOMPLETE="GOOGLE_PLACE_AUTOCOMPLETE",o.GOOGLE_SEARCH="GOOGLE_SEARCH",o.GOOGLE_TRANSLATE="GOOGLE_TRANSLATE",o.GPT_OSS="GPT_OSS",o.GPT_VISION="GPT_VISION",o.GROK_4_0709="GROK_4_0709",o.GROQ="GROQ",o.IMAGE_SEARCH="IMAGE_SEARCH",o.LOCAL_BUSINESS_DATA="LOCAL_BUSINESS_DATA",o.MOVIES_AND_TV_SERIES="MOVIES_AND_TV_SERIES",o.OPENAI_O3="OPENAI_O3",o.PDF_GENERATION="PDF_GENERATION",o.PLAGIARISM_CHECKER="PLAGIARISM_CHECKER",o.PRODUCT_SEARCH="PRODUCT_SEARCH",o.QR_CODE="QR_CODE",o.REKA="REKA",o.SALES_TAX="SALES_TAX",o.SCAN_DOCUMENT="SCAN_DOCUMENT",o.SEO_KEYWORD_RESEARCH="SEO_KEYWORD_RESEARCH",o.STABLE_DIFFUSION="STABLE_DIFFUSION",o.STABLE_DIFFUSION_V3="STABLE_DIFFUSION_V3",o.TEXT_MODERATION="TEXT_MODERATION",o.TEXT_TO_SPEECH="TEXT_TO_SPEECH",o.TRANSCRIBE_AUDIO="TRANSCRIBE_AUDIO",o.TRANSCRIBE_AUDIO_2="TRANSCRIBE_AUDIO_2",o.WEATHER_BY_CITY="WEATHER_BY_CITY",o.WEB_SCRAPING="WEB_SCRAPING",o))(S||{});var y=(a=>(a.ADD_AUTH_AND_CONTINUE_CHAT="ADD_AUTH_AND_CONTINUE_CHAT",a.AGENT_AUTO_SELECT_INTEGRATION="AGENT_AUTO_SELECT_INTEGRATION",a.AUTO_SELECT_CHAT="AUTO_SELECT_CHAT",a.CHANGE_DESIGN_SYSTEM_AND_CONTINUE_CHAT="CHANGE_DESIGN_SYSTEM_AND_CONTINUE_CHAT",a.CHAT="CHAT",a.DEEP_DEBUGGING="DEEP_DEBUGGING",a.DELEGATE_TO_SUB_AGENT="DELEGATE_TO_SUB_AGENT",a.DISCUSSION="DISCUSSION",a.DUPLICATE="DUPLICATE",a.EXPERT_SWE="EXPERT_SWE",a.EXTERNAL_UPDATE="EXTERNAL_UPDATE",a.FORK="FORK",a.FORK_MODULE_WITH_CHAT_MESSAGE="FORK_MODULE_WITH_CHAT_MESSAGE",a.FORK_WITH_CHAT_MESSAGE="FORK_WITH_CHAT_MESSAGE",a.INTRODUCE_BUG="INTRODUCE_BUG",a.JUDGE_DEEP_DEBUGGING="JUDGE_DEEP_DEBUGGING",a.ORCHESTRATOR="ORCHESTRATOR",a.PROJECT_APPLY_COMMAND="PROJECT_APPLY_COMMAND",a.PROJECT_CHANGE_AUTH="PROJECT_CHANGE_AUTH",a.PROJECT_CODE_CHANGE="PROJECT_CODE_CHANGE",a.PROJECT_CREATE_COMPONENT_FROM_PROMPT="PROJECT_CREATE_COMPONENT_FROM_PROMPT",a.PROJECT_CREATE_COMPONENT_FROM_PROMPT_AND_CODE="PROJECT_CREATE_COMPONENT_FROM_PROMPT_AND_CODE",a.PROJECT_DELETION="PROJECT_DELETION",a.PROJECT_DUPLICATE_MODULE="PROJECT_DUPLICATE_MODULE",a.PROJECT_DUPLICATE_PROJECT_GROUP="PROJECT_DUPLICATE_PROJECT_GROUP",a.PROJECT_FIX_CHANGE="PROJECT_FIX_CHANGE",a.PROJECT_FORK_FROM_PROJECT_GROUP="PROJECT_FORK_FROM_PROJECT_GROUP",a.PROJECT_FORK_FROM_PROJECT_GROUP_TEMPLATE="PROJECT_FORK_FROM_PROJECT_GROUP_TEMPLATE",a.PROJECT_IMAGE_CHANGE="PROJECT_IMAGE_CHANGE",a.PROJECT_INITIAL="PROJECT_INITIAL",a.PROJECT_LINK_CHANGE="PROJECT_LINK_CHANGE",a.PROJECT_REBUILD_FROM_LATEST="PROJECT_REBUILD_FROM_LATEST",a.PROJECT_RESTORE="PROJECT_RESTORE",a.PROJECT_TEXT_CHANGE="PROJECT_TEXT_CHANGE",a.PROJECT_UI_PATCH="PROJECT_UI_PATCH",a.PROJECT_UPDATE_REFERENCES="PROJECT_UPDATE_REFERENCES",a.PUBLISH_PROJECT_AND_CONTINUE_CHAT="PUBLISH_PROJECT_AND_CONTINUE_CHAT",a.RESTORE="RESTORE",a.SAVE_SECRET_AND_CONTINUE_CHAT="SAVE_SECRET_AND_CONTINUE_CHAT",a.SELECT_INTEGRATION_WITH_CHAT_MESSAGE="SELECT_INTEGRATION_WITH_CHAT_MESSAGE",a.UPDATE_AUTH_SETTINGS="UPDATE_AUTH_SETTINGS",a.UPDATE_EXPO_SETTINGS="UPDATE_EXPO_SETTINGS",a))(y||{});var I=p.gql`
	fragment CreateProjectFragment on Project {
		id
		name
		slug
		lastEditedAt
		moduleType
		organization {
			id
		}
		projectGroup {
			id
			domains {
				id
				domain
			}
			publishState
			slug
		}
		previewURL(aspectRatio: 1.0)
		pathSegment
		filePath
	}
`,d=p.gql`
	fragment RevisionForProject on ProjectRevision {
		id
		action
		content(replaceFunctions: $replaceFunctions)
		context
		prompt
		cleanedPrompt
		createdAt
		status
		previewURL
		componentNameMapping
		dataStoreNameMapping
		functionNameMapping
		assetNameMapping
		needsAuth
		generationAttempt
		filePath
		routingBehavior
	}
`,u=p.gql`
	fragment ProjectGroupRevisionForChat on ProjectGroupRevision {
		id
		response
		status
		action
		previewURL
		chat {
			id
			content
		}
		createdAt
		thread {
			id
			updatedAt
		}
	}
`,s=p.gql`
	fragment ThreadForProjectGroupChat on ProjectGroupRevisionThread {
		id
		updatedAt
		title
		latestRevision {
			...ProjectGroupRevisionForChat
		}
	}
	${u}
`,A=p.gql`
	mutation CreateNewFile($input: CreateFileInput!) {
		createFile(input: $input) {
			success
			project {
				...CreateProjectFragment
			}
		}
	}
	${I}
`;function V(e){let t={...i,...e};return p__namespace.useMutation(A,t)}var g=p.gql`
	mutation CreateProjectRevision(
		$projectId: ID!
		$content: String!
		$action: ProjectRevisionAction
		$context: String
		$replaceFunctions: Boolean
	) {
		createRevision(
			input: {
				projectId: $projectId
				content: $content
				context: $context
				action: $action
			}
		) {
			success
			projectRevision {
				...RevisionForProject
			}
		}
	}
	${d}
`;function z(e){let t={...i,...e};return p__namespace.useMutation(g,t)}var P=p.gql`
	query FilesForProjectGroup($id: ID!) {
		projectGroupById(id: $id) {
			id
			files {
				id
				latestRevision {
					id
					filePath
				}
			}
			projectFiles {
				projectId
				path
				content
			}
		}
	}
`;function W(e){let t={...i,...e};return p__namespace.useQuery(P,t)}var w=p.gql`
	subscription GetModulesDeletedForProjectGroup($id: ID!) {
		projectGroupModulesDeleted(id: $id) {
			id
			moduleIds
		}
	}
`;var K=p.gql`
	subscription ProjectGroupRevisionsFinished($id: ID!) {
		projectGroupRevisionsFinished(id: $id) {
			id
			revisionIds
		}
	}
`;var _=p.gql`
	mutation RequestMagicLink($input: RequestMagicLinkInput!) {
		requestMagicLink(input: $input) {
			verificationRequested {
				requested
				expiresAt
			}
		}
	}
`;function Y(e){let t={...i,...e};return p__namespace.useMutation(_,t)}var R=p.gql`
	mutation SignInWithMagicLinkCode($input: SignInWithMagicLinkCodeInput!) {
		signInWithMagicLinkCode(input: $input) {
			accessToken
			refreshToken
		}
	}
`;function $(e){let t={...i,...e};return p__namespace.useMutation(R,t)}var M=p.gql`
	mutation SignUpWithMagicLinkOrRequestMagicLink(
		$input: SignUpWithMagicLinkOrRequestMagicLinkInput!
	) {
		signUpWithMagicLinkOrRequestMagicLink(input: $input) {
			success
		}
	}
`;function J(e){let t={...i,...e};return p__namespace.useMutation(M,t)}var G=p.gql`
	query GetProjectGroups(
		$first: Int
		$after: String
		$last: Int
		$before: String
		$organizationId: ID!
		$input: ProjectGroupsInput!
	) {
		organizationById(id: $organizationId) {
			id
			creditBalance
			plan
		}
		projectGroups(
			first: $first
			after: $after
			last: $last
			before: $before
			input: $input
		) {
			pageInfo {
				hasNextPage
				hasPreviousPage
				endCursor
				startCursor
			}
			edges {
				node {
					id
					name
					updatedAt
					isFileBased
					slug
					organizationId
					lastRevisionWithExpoBundle {
						id
					}
					latestRevision {
						expoSettings {
							appIconUrl
						}
					}
					mobileDevServerSession {
						qrCode
					}
					isMobile
				}
			}
		}
	}
`;function q(e){let t={...i,...e};return p__namespace.useQuery(G,t)}var m=p.gql`
	mutation DeleteProjectGroup($input: DeleteProjectGroupInput!) {
		deleteProjectGroup(input: $input) {
			success
		}
	}
`;function X(e){let t={...i,...e};return p__namespace.useMutation(m,t)}var E=p.gql`
	mutation RenameProjectGroup($input: RenameProjectGroupInput!) {
		renameProjectGroup(input: $input) {
			success
			projectGroup {
				id
				name
			}
		}
	}
`;function Z(e){let t={...i,...e};return p__namespace.useMutation(E,t)}var D=p.gql`
	mutation DeleteUser($input: DeleteUserInput!) {
		deleteUser(input: $input) {
			success
		}
	}
`;function ee(e){let t={...i,...e};return p__namespace.useMutation(D,t)}var C=p.gql`
	query GetAggregatedUsageByOrganizationId(
		$id: ID!
		$startDate: DateTimeISO!
		$endDate: DateTimeISO!
	) {
		organizationById(id: $id) {
			id
			aggregatedIntegrationCreditUsageBySubType(
				startDate: $startDate
				endDate: $endDate
			) {
				total
				usages {
					amount
					average
					type
					subType
				}
			}
			aggregatedCreditUsageByType(startDate: $startDate, endDate: $endDate) {
				total
				usages {
					amount
					average
					type
					subType
				}
			}
		}
	}
`;function te(e){let t={...i,...e};return p__namespace.useQuery(C,t)}var O=p.gql`
	mutation CreateProjectGroup($input: CreateProjectGroupInput!) {
		createProjectGroup(input: $input) {
			success
			projectGroup {
				id
				name
			}
			project {
				id
			}
		}
	}
`;function oe(e){let t={...i,...e};return p__namespace.useMutation(O,t)}var b=p.gql`
	mutation AddSecret($input: AddSecretInput!) {
		addSecret(input: $input) {
			success
			secret {
				id
				displayName
				value
				environment
				createdAt
				updatedAt
			}
			projectGroup {
				id
			}
		}
	}
`;function ae(e){let t={...i,...e};return p__namespace.useMutation(b,t)}var T=p.gql`
	mutation CancelLatestInflightRevision(
		$input: CancelLatestInflightRevisionInput!
	) {
		cancelLatestInflightRevision(input: $input) {
			success
		}
	}
`;function re(e){let t={...i,...e};return p__namespace.useMutation(T,t)}var j=p.gql`
	mutation CancelLatestInflightRevisionForThread(
		$input: CancelLatestInflightRevisionForThreadInput!
	) {
		cancelLatestInflightRevisionForThread(input: $input) {
			success
		}
	}
`;function ne(e){let t={...i,...e};return p__namespace.useMutation(j,t)}var pe=p.gql`
	query ChatUsageByOrganizationId($organizationId: ID!) {
		organizationById(id: $organizationId) {
			id
			plan
			chatLimits {
				window
				count
				limit
				hasReachedLimit
			}
			hasReachedChatCreditsLimit
			subscriptionStartDate
		}
	}
`;var N=p.gql`
	mutation CreateAsset($input: CreateAssetInput!) {
		createAsset(input: $input) {
			success
			asset {
				id
				url
				userDefinedName
				mimeType
				createdAt
				updatedAt
			}
		}
	}
`;function ie(e){let t={...i,...e};return p__namespace.useMutation(N,t)}var U=p.gql`
	mutation DeleteAsset($input: DeleteAssetInput!) {
		deleteAsset(input: $input) {
			success
		}
	}
`;function ue(e){let t={...i,...e};return p__namespace.useMutation(U,t)}var x=p.gql`
	mutation DeleteSecret($input: DeleteSecretInput!) {
		deleteSecret(input: $input) {
			success
		}
	}
`;function se(e){let t={...i,...e};return p__namespace.useMutation(x,t)}var L=p.gql`
	mutation GenerateProjectGroupRevisionFromChat(
		$input: GenerateProjectGroupRevisionFromChatInput!
	) {
		generateProjectGroupRevisionFromChat(input: $input) {
			success
			projectGroupRevision {
				id
				...ProjectGroupRevisionForChat
				thread {
					...ThreadForProjectGroupChat
				}
			}
			askForUserProfileInfo
		}
	}
	${u}
	${s}
`;function le(e){let t={...i,...e};return p__namespace.useMutation(L,t)}var v=p.gql`
	query GetAssets($first: Int, $after: String, $input: AssetInput!) {
		assets(first: $first, after: $after, input: $input) {
			edges {
				node {
					id
					url
					userDefinedName
					mimeType
					createdAt
					updatedAt
				}
			}
			pageInfo {
				hasNextPage
				endCursor
			}
			totalCount
		}
	}
`;function ce(e){let t={...i,...e};return p__namespace.useQuery(v,t)}p.gql`
	query GetMobileDevServerSession($projectGroupId: ID!) {
		projectGroupById(id: $projectGroupId) {
			id
			mobileDevServerSession {
				id
				qrCode
				builderUrl
			}
		}
	}
`;var ye=p.gql`
	subscription GetModulesCreatedForProjectGroup($id: ID!) {
		projectGroupModulesCreated(id: $id) {
			id
			moduleIds
		}
	}
`;p.gql`
	query GetProjectFilePathsForGroup($groupId: ID!) {
		projects(input: { groupId: $groupId }) {
			edges {
				node {
					id
					name
					filePath
				}
			}
		}
	}
`;var F=p.gql`
	query GetProjectGroupSecrets(
		$projectGroupId: ID!
		$first: Int
		$after: String
		$environment: CreateEnvironment
	) {
		projectGroupById(id: $projectGroupId) {
			id
			secrets(first: $first, after: $after, environment: $environment) {
				totalCount
				count
				pageInfo {
					hasNextPage
					hasPreviousPage
					endCursor
					startCursor
				}
				nodes {
					id
					displayName
					value
					environment
					createdAt
					updatedAt
				}
			}
		}
	}
`;function de(e){let t={...i,...e};return p__namespace.useQuery(F,t)}p.gql`
	query ProjectGroupHasModules($projectGroupId: ID!) {
		projectGroupById(id: $projectGroupId) {
			id
			hasModules
			hasRevision
		}
	}
`;var h=p.gql`
	query GetProjectGroupRevisionsForChat(
		$projectGroupId: ID!
		$threadId: ID
		$first: Int
		$after: String
	) {
		projectGroupById(id: $projectGroupId) {
			id
			latestRevision: latestRevisionForThread(threadId: $threadId) {
				id
				status
				createdAt
				thread {
					id
				}
			}
			revisions(
				order: [{ createdAt: "DESC" }]
				first: $first
				after: $after
				input: {
					threadId: $threadId
					actions: [
						CHAT
						AUTO_SELECT_CHAT
						DEEP_DEBUGGING
						EXPERT_SWE
						DISCUSSION
						FORK_WITH_CHAT_MESSAGE
						FORK_MODULE_WITH_CHAT_MESSAGE
						SELECT_INTEGRATION_WITH_CHAT_MESSAGE
						PUBLISH_PROJECT_AND_CONTINUE_CHAT
						CHANGE_DESIGN_SYSTEM_AND_CONTINUE_CHAT
						SAVE_SECRET_AND_CONTINUE_CHAT
						ADD_AUTH_AND_CONTINUE_CHAT
						RESTORE
					]
				}
			) {
				totalCount
				pageInfo {
					hasNextPage
					hasPreviousPage
					endCursor
					startCursor
				}
				edges {
					node {
						id
						response
						status
						action
						previewURL
						chat {
							content
						}
						createdAt
						thread {
							id
							updatedAt
						}
					}
				}
			}
		}
	}
`;function ge(e){let t={...i,...e};return p__namespace.useQuery(h,t)}var Pe=p.gql`
	subscription GetProjectGroupRevisionContentUpdate($id: ID!) {
		projectGroupRevisionContentUpdate(id: $id) {
			id
			response
			status
		}
	}
`;var B=p.gql`
	query ProjectGroupRevisionForChatById($id: ID!) {
		projectGroupRevisionById(id: $id) {
			...ProjectGroupRevisionForChat
		}
	}
	${u}
`;function _e(e){let t={...i,...e};return p__namespace.useLazyQuery(B,t)}var Re=p.gql`
	subscription ProjectGroupRevisionsCreated($id: ID!) {
		projectGroupRevisionsCreated(id: $id) {
			id
			revisionIds
		}
	}
`;var Me=p.gql`
	query GetThreadsForProjectGroup(
		$input: ProjectGroupRevisionThreadsByProjectGroupIdInput!
		$projectGroupId: ID!
		$after: String
	) {
		projectGroupById(id: $projectGroupId) {
			id
			latestThread {
				id
			}
		}
		projectGroupRevisionThreadsByProjectGroupId(input: $input, after: $after) {
			pageInfo {
				hasNextPage
				endCursor
			}
			edges {
				node {
					...ThreadForProjectGroupChat
				}
			}
		}
	}
	${s}
`;var k=p.gql`
	mutation BootDevServer($input: BootDevServerInput!) {
		bootDevServer(input: $input) {
			success
			devServerSession {
				id
				status
				webUrl
				mobileUrl
			}
		}
	}
`;function Ge(e){let t={...i,...e};return p__namespace.useMutation(k,t)}var Q=p.gql`
	query GetOrganizations {
		organizations(order: { createdAt: "ASC" }) {
			edges {
				node {
					id
					plan
					planId
					canUseMax
					name
					logoUrl
					hasReachedChatCreditsLimit
					creditBalance
					revenueCatUserId
					planCredits
					subscriptionStartDate
				}
			}
		}
	}
`;function me(e){let t={...i,...e};return p__namespace.useQuery(Q,t)}var l=p.gql`
	query Me {
		me {
			id
			profile {
				photoURL
			}
			email
			displayName
		}
	}
`;function Ee(e){let t={...i,...e};return p__namespace.useQuery(l,t)}function De(e){let t={...i,...e};return p__namespace.useLazyQuery(l,t)}var f=p.gql`
	subscription SubscribeToDevServerUpdates($id: String!) {
		devServerUpdate(id: $id) {
			id
			webUrl
			mobileUrl
			status
		}
	}
`;function Ce(e){let t={...i,...e};return p__namespace.useSubscription(f,t)}exports.A=ye;exports.B=de;exports.C=h;exports.D=ge;exports.E=Pe;exports.F=_e;exports.G=Re;exports.H=Me;exports.I=Ge;exports.J=me;exports.K=Ee;exports.L=De;exports.M=Ce;exports.a=c;exports.b=S;exports.c=y;exports.d=V;exports.e=z;exports.f=W;exports.g=w;exports.h=K;exports.i=Y;exports.j=$;exports.k=J;exports.l=q;exports.m=X;exports.n=Z;exports.o=ee;exports.p=te;exports.q=oe;exports.r=ae;exports.s=re;exports.t=ne;exports.u=pe;exports.v=ie;exports.w=ue;exports.x=se;exports.y=le;exports.z=ce;
