// react-native-screen-transitions/lib/commonjs/shared/hooks/animation/use-screen-animation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._useScreenAnimation = _useScreenAnimation;
exports.useScreenAnimation = useScreenAnimation;
var _react = require("react");
var _reactNative = require("react-native");
var _reactNativeReanimated = require("react-native-reanimated");
var _reactNativeSafeAreaContext = require("react-native-safe-area-context");
var _constants = require("../../constants");
var _keys = require("../../providers/screen/keys.provider");
var _animation = require("../../stores/animation.store");
var _gesture = require("../../stores/gesture.store");
var _derivations = require("../../utils/animation/derivations");
var _bounds = require("../../utils/bounds");
var _useStack = require("../navigation/use-stack");
const createScreenTransitionState = (route, meta) => ({
  progress: 0,
  closing: 0,
  animating: 0,
  entering: 1,
  gesture: {
    x: 0,
    y: 0,
    normalizedX: 0,
    normalizedY: 0,
    isDismissing: 0,
    isDragging: 0,
    direction: null
  },
  route,
  meta
});
const unwrapInto = s => {
  "worklet";

  const out = s.unwrapped;
  out.progress = s.progress.value;
  out.closing = s.closing.value;
  out.entering = s.entering.value;
  out.animating = s.animating.value;
  out.gesture.x = s.gesture.x.value;
  out.gesture.y = s.gesture.y.value;
  out.gesture.normalizedX = s.gesture.normalizedX.value;
  out.gesture.normalizedY = s.gesture.normalizedY.value;
  out.gesture.isDismissing = s.gesture.isDismissing.value;
  out.gesture.isDragging = s.gesture.isDragging.value;
  out.gesture.direction = s.gesture.direction.value;
  out.meta = s.meta;
  return out;
};
const useBuildScreenTransitionState = descriptor => {
  const key = descriptor?.route.key;
  const meta = descriptor?.options?.meta;
  return (0, _react.useMemo)(() => {
    if (!key) return undefined;
    return {
      progress: _animation.AnimationStore.getAnimation(key, "progress"),
      closing: _animation.AnimationStore.getAnimation(key, "closing"),
      entering: _animation.AnimationStore.getAnimation(key, "entering"),
      animating: _animation.AnimationStore.getAnimation(key, "animating"),
      gesture: _gesture.GestureStore.getRouteGestures(key),
      route: descriptor.route,
      meta,
      unwrapped: createScreenTransitionState(descriptor.route, meta)
    };
  }, [key, descriptor?.route, meta]);
};
const hasTransitionsEnabled = (options, alwaysOn) => {
  "worklet";

  if (alwaysOn) return true;
  return !!options?.enableTransitions;
};
function _useScreenAnimation() {
  const windowDimensions = (0, _reactNative.useWindowDimensions)();
  const dimensions = windowDimensions;
  const insets = (0, _reactNativeSafeAreaContext.useSafeAreaInsets)();
  const {
    flags,
    stackProgress: rootStackProgress,
    routeKeys
  } = (0, _useStack.useStack)();
  const transitionsAlwaysOn = flags.TRANSITIONS_ALWAYS_ON;
  const {
    current: currentDescriptor,
    next: nextDescriptor,
    previous: previousDescriptor
  } = (0, _keys.useKeys)();
  const currentAnimation = useBuildScreenTransitionState(currentDescriptor);
  const nextAnimation = useBuildScreenTransitionState(nextDescriptor);
  const prevAnimation = useBuildScreenTransitionState(previousDescriptor);
  const currentRouteKey = currentDescriptor?.route?.key;
  const currentIndex = routeKeys.indexOf(currentRouteKey);
  const screenInterpolatorProps = (0, _reactNativeReanimated.useDerivedValue)(() => {
    "worklet";

    const previous = prevAnimation ? unwrapInto(prevAnimation) : undefined;
    const next = nextAnimation && hasTransitionsEnabled(nextDescriptor?.options, transitionsAlwaysOn) ? unwrapInto(nextAnimation) : undefined;
    const current = currentAnimation ? unwrapInto(currentAnimation) : _constants.DEFAULT_SCREEN_TRANSITION_STATE;
    const {
      progress,
      ...helpers
    } = (0, _derivations.derivations)({
      previous,
      current,
      next
    });

    // Compute relative stack progress: total - currentIndex
    // This gives us the sum of progress values from current screen onwards
    // Falls back to current progress if index is invalid
    const stackProgress = currentIndex >= 0 ? rootStackProgress.value - currentIndex : progress;
    return {
      layouts: {
        screen: dimensions
      },
      insets,
      previous,
      current,
      next,
      progress,
      stackProgress,
      ...helpers
    };
  });
  const nextInterpolator = nextDescriptor?.options.screenStyleInterpolator;
  const currentInterpolator = currentDescriptor?.options.screenStyleInterpolator;
  return {
    screenInterpolatorProps,
    nextInterpolator,
    currentInterpolator
  };
}
function useScreenAnimation() {
  const {
    screenInterpolatorProps
  } = _useScreenAnimation();
  return (0, _reactNativeReanimated.useDerivedValue)(() => {
    const props = screenInterpolatorProps.value;
    return {
      ...props,
      bounds: (0, _bounds.createBounds)(props)
    };
  });
}
