// react-native-screen-transitions/lib/commonjs/shared/hooks/use-stable-callback-value.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useStableCallbackValue;
var _react = require("react");
var _reactNativeReanimated = require("react-native-reanimated");
/**
 * Credit:
 * https://github.com/MatiPl01/react-native-sortables/blob/main/packages/react-native-sortables/src/integrations/reanimated/hooks/useStableCallbackValue.ts
 */

function useMutableValue(initialValue) {
  const [mutable] = (0, _react.useState)(() => (0, _reactNativeReanimated.makeMutable)(initialValue));
  (0, _react.useEffect)(() => {
    return () => {
      (0, _reactNativeReanimated.cancelAnimation)(mutable);
    };
  }, [mutable]);
  return mutable;
}

// We cannot store a function as a SharedValue because reanimated will treat
// it as an animation and will try to execute the animation when assigned
// to the SharedValue. Since we want the function to be treated as a value,
// we wrap it in an object and store the object in the SharedValue.

const wrap = callback => {
  if ((0, _reactNativeReanimated.isWorkletFunction)(callback)) {
    return {
      call: callback
    };
  }
  return {
    call: (...args) => {
      "worklet";

      (0, _reactNativeReanimated.runOnJS)(callback)(...args);
    }
  };
};

/** Creates a stable worklet callback that can be called from the UI runtime
 * @param callback The JavaScript or worklet function to be called
 * @returns A worklet function that can be safely called from the UI thread
 * @default Behavior:
 * - If passed a regular JS function, calls it on the JS thread using runOnJS
 * - If passed a worklet function, calls it directly on the UI thread
 * @important The returned function maintains a stable reference and properly handles
 * thread execution based on the input callback type
 */
function useStableCallbackValue(callback) {
  const stableCallback = useMutableValue(null);
  (0, _react.useEffect)(() => {
    if (callback) {
      stableCallback.value = wrap(callback);
    } else {
      stableCallback.value = null;
    }
  }, [callback, stableCallback]);
  return (0, _react.useCallback)((...args) => {
    "worklet";

    stableCallback.value?.call(...args);
  }, [stableCallback]);
}
