// react-native-screen-transitions/lib/commonjs/shared/utils/bounds/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createBounds = void 0;
var _reactNativeReanimated = require("react-native-reanimated");
var _constants = require("../../constants");
var _bounds = require("../../stores/bounds.store");
var _geometry = require("./helpers/geometry");
var _interpolateStyle = require("./helpers/interpolate-style");
var _styleComposers = require("./helpers/style-composers");
const resolveBounds = props => {
  "worklet";

  const entering = !props.next;
  const fullscreen = (0, _constants.FULLSCREEN_DIMENSIONS)(props.dimensions);
  const isFullscreenTarget = props.computeOptions.target === "fullscreen";

  // Try exact match first (strict matching for nested stacks)
  let link = _bounds.BoundStore.getActiveLink(props.id, props.current?.route.key);

  // For fullscreen target, fall back to most recent link for this tag
  // (destination screen might not have a matching element)
  if (!link && isFullscreenTarget) {
    link = _bounds.BoundStore.getActiveLink(props.id); // No screenKey = get most recent
  }
  if (!link || !link.source) {
    return {
      start: null,
      end: null,
      entering
    };
  }

  // For fullscreen target, destination element is not required
  if (!isFullscreenTarget && !link.destination) {
    return {
      start: null,
      end: null,
      entering
    };
  }
  const {
    destination,
    source
  } = link;
  const start = source.bounds;
  let end = destination?.bounds ?? fullscreen;
  if (isFullscreenTarget) {
    end = fullscreen;
  }
  const customTarget = props.computeOptions.target;
  if (typeof customTarget === "object") {
    end = customTarget;
  }
  return {
    start,
    end,
    entering
  };
};
const computeBoundStyles = ({
  id,
  previous,
  current,
  next,
  progress,
  dimensions
}, computeOptions = {
  id: "bound-id"
}) => {
  "worklet";

  if (!id) {
    if (computeOptions.raw) {
      return _constants.EMPTY_BOUND_HELPER_RESULT_RAW;
    }
    return _constants.EMPTY_BOUND_HELPER_RESULT;
  }
  const {
    start,
    end,
    entering
  } = resolveBounds({
    id,
    previous,
    current,
    next,
    computeOptions,
    dimensions
  });
  if (!start || !end) {
    if (computeOptions.raw) {
      return _constants.EMPTY_BOUND_HELPER_RESULT_RAW;
    }
    return _constants.EMPTY_BOUND_HELPER_RESULT;
  }
  const ranges = entering ? _constants.ENTER_RANGE : _constants.EXIT_RANGE;
  if (computeOptions.method === "content") {
    const geometry = (0, _geometry.computeContentTransformGeometry)({
      start,
      end,
      entering,
      dimensions,
      anchor: computeOptions.anchor,
      scaleMode: computeOptions.scaleMode
    });
    return (0, _styleComposers.composeContentStyle)({
      start,
      progress,
      ranges,
      end,
      geometry,
      computeOptions
    });
  }
  const geometry = (0, _geometry.computeRelativeGeometry)({
    start,
    end,
    entering,
    anchor: computeOptions.anchor,
    scaleMode: computeOptions.scaleMode
  });
  const common = {
    start,
    end,
    progress,
    ranges,
    geometry,
    computeOptions
  };
  const isSize = computeOptions.method === "size";
  const isAbs = computeOptions.space === "absolute";
  return isSize ? isAbs ? (0, _styleComposers.composeSizeAbsolute)(common) : (0, _styleComposers.composeSizeRelative)(common) : isAbs ? (0, _styleComposers.composeTransformAbsolute)(common) : (0, _styleComposers.composeTransformRelative)(common);
};
const createBounds = props => {
  "worklet";

  const boundsFunction = params => {
    "worklet";

    const id = params?.id;
    return computeBoundStyles({
      id,
      previous: props.previous,
      current: props.current,
      next: props.next,
      progress: props.progress,
      dimensions: props.layouts.screen
    }, params);
  };
  const getSnapshot = (tag, key) => {
    "worklet";

    return _bounds.BoundStore.getSnapshot(tag, key);
  };
  const getLink = tag => {
    "worklet";

    const link = _bounds.BoundStore.getActiveLink(tag, props.current?.route.key);
    if (!link) return null;
    return {
      source: link.source ? {
        bounds: link.source.bounds,
        styles: link.source.styles
      } : null,
      destination: link.destination ? {
        bounds: link.destination.bounds,
        styles: link.destination.styles
      } : null
    };
  };
  const interpolateStyle = (tag, property, fallback) => {
    "worklet";

    const link = getLink(tag);
    const entering = !props.next;
    return (0, _interpolateStyle.interpolateLinkStyle)(link, property, props.progress, entering, {
      fallback
    });
  };
  const interpolateBounds = (tag, property, fallbackOrTargetKey, fallback) => {
    "worklet";

    const entering = !props.next;
    const range = entering ? _constants.ENTER_RANGE : _constants.EXIT_RANGE;

    // If third param is a string, it's a targetKey (snapshot approach)
    if (typeof fallbackOrTargetKey === "string") {
      const targetKey = fallbackOrTargetKey;
      const currentKey = props.current?.route?.key;
      const fb = fallback ?? 0;
      const currentSnapshot = currentKey ? _bounds.BoundStore.getSnapshot(tag, currentKey) : null;
      const targetSnapshot = _bounds.BoundStore.getSnapshot(tag, targetKey);
      const currentValue = currentSnapshot?.bounds?.[property] ?? fb;
      const targetValue = targetSnapshot?.bounds?.[property] ?? fb;
      return (0, _reactNativeReanimated.interpolate)(props.progress, range, [targetValue, currentValue], "clamp");
    }

    // Otherwise, use link approach (existing behavior)
    const link = getLink(tag);
    const fb = fallbackOrTargetKey ?? 0;
    const sourceValue = link?.source?.bounds?.[property] ?? fb;
    const destValue = link?.destination?.bounds?.[property] ?? fb;
    return (0, _reactNativeReanimated.interpolate)(props.progress, range, [sourceValue, destValue], "clamp");
  };
  return Object.assign(boundsFunction, {
    getSnapshot,
    getLink,
    interpolateStyle,
    interpolateBounds
  });
};
exports.createBounds = createBounds;
