// react-native-screen-transitions/lib/commonjs/shared/providers/stack/core.provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStackCoreContext = void 0;
exports.withStackCore = withStackCore;
var _elements = require("@react-navigation/elements");
var _reactNative = require("react-native");
var _reactNativeGestureHandler = require("react-native-gesture-handler");
var _stack = require("../../types/stack.types");
var _createProvider = _interopRequireDefault(require("../../utils/create-provider"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const {
  StackCoreProvider: InternalStackCoreProvider,
  useStackCoreContext
} = (0, _createProvider.default)("StackCore", {
  guarded: true
})(({
  config,
  children
}) => {
  const {
    TRANSITIONS_ALWAYS_ON = false,
    DISABLE_NATIVE_SCREENS = false,
    STACK_TYPE = _stack.StackType.BLANK
  } = config;
  return {
    value: {
      flags: {
        TRANSITIONS_ALWAYS_ON,
        STACK_TYPE,
        DISABLE_NATIVE_SCREENS
      }
    },
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeGestureHandler.GestureHandlerRootView, {
      style: styles.container,
      pointerEvents: STACK_TYPE === _stack.StackType.COMPONENT ? "box-none" : undefined,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_elements.SafeAreaProviderCompat, {
        children: children
      })
    })
  };
});

/**
 * HOC that wraps a component with the StackCore provider.
 * Just a simple open gate
 */
exports.useStackCoreContext = useStackCoreContext;
function withStackCore(defaultConfig, Component) {
  return function StackCoreWrapper({
    DISABLE_NATIVE_SCREENS,
    TRANSITIONS_ALWAYS_ON,
    STACK_TYPE,
    ...props
  }) {
    // User props first, then defaultConfig overrides (config takes priority)
    const config = {
      ...(DISABLE_NATIVE_SCREENS !== undefined && {
        DISABLE_NATIVE_SCREENS
      }),
      ...(TRANSITIONS_ALWAYS_ON !== undefined && {
        TRANSITIONS_ALWAYS_ON
      }),
      ...(STACK_TYPE !== undefined && {
        STACK_TYPE
      }),
      ...defaultConfig
    };
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(InternalStackCoreProvider, {
      config: config,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, {
        ...props
      })
    });
  };
}
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  }
});
