// react-native-screen-transitions/lib/commonjs/shared/providers/stack/direct.provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDirectStackContext = useDirectStackContext;
exports.withDirectStack = withDirectStack;
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _reactNativeReanimated = require("react-native-reanimated");
var _useStack = require("../../hooks/navigation/use-stack");
var _animation = require("../../stores/animation.store");
var _history = require("../../stores/history.store");
var _core = require("./core.provider");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const DirectStackContext = /*#__PURE__*/React.createContext(null);
DirectStackContext.displayName = "DirectStack";
function useDirectStackContext() {
  const context = React.useContext(DirectStackContext);
  if (!context) {
    throw new Error("useDirectStackContext must be used within DirectStackProvider");
  }
  return context;
}

/**
 * Internal hook that computes all lifecycle values.
 */
function useDirectStackValue(props) {
  const {
    state,
    navigation,
    descriptors,
    describe
  } = props;
  const {
    flags
  } = (0, _core.useStackCoreContext)();
  const preloadedDescriptors = (0, _react.useMemo)(() => {
    return state.preloadedRoutes.reduce((acc, route) => {
      acc[route.key] = acc[route.key] || describe(route, true);
      return acc;
    }, {});
  }, [state.preloadedRoutes, describe]);
  const {
    scenes,
    shouldShowFloatOverlay,
    routeKeys,
    allRoutes,
    allDescriptors
  } = (0, _react.useMemo)(() => {
    const allRoutes = state.routes.concat(state.preloadedRoutes);
    const scenes = [];
    const routeKeys = [];
    const allDescriptors = {
      ...preloadedDescriptors,
      ...descriptors
    };
    let shouldShowFloatOverlay = false;
    for (const route of allRoutes) {
      const descriptor = allDescriptors[route.key];
      const isPreloaded = preloadedDescriptors[route.key] !== undefined && descriptors[route.key] === undefined;
      scenes.push({
        route,
        descriptor,
        isPreloaded
      });
      routeKeys.push(route.key);
      if (!shouldShowFloatOverlay && descriptor) {
        const options = descriptor.options;
        if (options?.enableTransitions === true && options?.overlayMode === "float" && options?.overlayShown === true) {
          shouldShowFloatOverlay = true;
        }
      }
    }
    return {
      scenes,
      shouldShowFloatOverlay,
      routeKeys,
      allRoutes,
      allDescriptors
    };
  }, [state.routes, state.preloadedRoutes, descriptors, preloadedDescriptors]);

  // Get animation store maps for all routes
  const animationMaps = (0, _react.useMemo)(() => allRoutes.map(route => _animation.AnimationStore.getAll(route.key)), [allRoutes]);
  const stackProgress = (0, _reactNativeReanimated.useDerivedValue)(() => {
    "worklet";

    let total = 0;
    for (let i = 0; i < animationMaps.length; i++) {
      total += animationMaps[i].progress.value;
    }
    return total;
  });
  const optimisticFocusedIndex = (0, _reactNativeReanimated.useDerivedValue)(() => {
    "worklet";

    const currentIndex = animationMaps.length - 1;
    let isAnyClosing = false;
    for (let i = 0; i < animationMaps.length; i++) {
      if (animationMaps[i].closing.value > 0) {
        isAnyClosing = true;
        break;
      }
    }
    return currentIndex - (isAnyClosing ? 1 : 0);
  });
  const focusedIndex = state.index;
  const stackContextValue = (0, _react.useMemo)(() => ({
    flags,
    routeKeys,
    routes: allRoutes,
    descriptors: allDescriptors,
    scenes,
    focusedIndex,
    stackProgress,
    optimisticFocusedIndex
  }), [routeKeys, allRoutes, allDescriptors, scenes, focusedIndex, stackProgress, optimisticFocusedIndex, flags]);

  // DirectStack context value
  const lifecycleValue = (0, _react.useMemo)(() => ({
    state,
    navigation,
    descriptors,
    preloadedDescriptors,
    scenes,
    focusedIndex,
    shouldShowFloatOverlay,
    stackProgress,
    optimisticFocusedIndex
  }), [state, navigation, descriptors, preloadedDescriptors, scenes, focusedIndex, shouldShowFloatOverlay, stackProgress, optimisticFocusedIndex]);
  return {
    ...lifecycleValue,
    stackContextValue
  };
}

/**
 * HOC that wraps component with DirectStack provider AND StackContext.
 * Used by native-stack which uses navigation state directly (no local route management).
 */
function withDirectStack(Component) {
  return function DirectStackProvider(props) {
    const navigatorKey = props.state.key;

    // Clean up history when navigator unmounts
    (0, _react.useEffect)(() => {
      return () => {
        _history.HistoryStore.clearNavigator(navigatorKey);
      };
    }, [navigatorKey]);
    const {
      stackContextValue,
      ...lifecycleValue
    } = useDirectStackValue(props);
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_useStack.StackContext.Provider, {
      value: stackContextValue,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DirectStackContext.Provider, {
        value: lifecycleValue,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, {
          ...lifecycleValue
        })
      })
    });
  };
}
