// react-native-screen-transitions/lib/commonjs/shared/providers/stack/managed.provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useManagedStackContext = useManagedStackContext;
exports.withManagedStack = withManagedStack;
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _reactNativeReanimated = require("react-native-reanimated");
var _useStack = require("../../hooks/navigation/use-stack");
var _animation = require("../../stores/animation.store");
var _history = require("../../stores/history.store");
var _core = require("./core.provider");
var _activeScreensLimit = require("./helpers/active-screens-limit");
var _useLocalRoutes = require("./helpers/use-local-routes");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * Props for managed stack - generic over descriptor and navigation types.
 * Defaults to base types for backward compatibility.
 */

/**
 * Context value for managed stack - generic over descriptor type.
 */

const ManagedStackContext = /*#__PURE__*/React.createContext(null);
ManagedStackContext.displayName = "ManagedStack";
function useManagedStackContext() {
  const context = React.useContext(ManagedStackContext);
  if (!context) {
    throw new Error("useManagedStackContext must be used within ManagedStackProvider");
  }
  return context;
}
function useManagedStackValue(props) {
  const {
    flags
  } = (0, _core.useStackCoreContext)();
  const {
    state,
    handleCloseRoute,
    closingRouteKeys
  } = (0, _useLocalRoutes.useLocalRoutes)(props);
  const {
    scenes,
    activeScreensLimit,
    shouldShowFloatOverlay,
    routeKeys
  } = (0, _react.useMemo)(() => {
    const scenes = [];
    const routeKeys = [];
    let shouldShowFloatOverlay = false;
    for (const route of state.routes) {
      const descriptor = state.descriptors[route.key];
      scenes.push({
        route,
        descriptor
      });
      routeKeys.push(route.key);
      if (!shouldShowFloatOverlay) {
        const options = descriptor?.options;
        shouldShowFloatOverlay = options?.overlayMode === "float" && options?.overlayShown === true;
      }
    }
    return {
      scenes,
      routeKeys,
      activeScreensLimit: (0, _activeScreensLimit.calculateActiveScreensLimit)(state.routes, state.descriptors),
      shouldShowFloatOverlay
    };
  }, [state.routes, state.descriptors]);

  // Get animation store maps for LOCAL routes (including closing routes)
  const animationMaps = (0, _react.useMemo)(() => state.routes.map(route => _animation.AnimationStore.getAll(route.key)), [state.routes]);

  // Aggregated stack progress from LOCAL routes (includes closing routes)
  const stackProgress = (0, _reactNativeReanimated.useDerivedValue)(() => {
    "worklet";

    let total = 0;
    for (let i = 0; i < animationMaps.length; i++) {
      total += animationMaps[i].progress.value;
    }
    return total;
  });

  // Optimistic focused index: accounts for closing screens
  const optimisticFocusedIndex = (0, _reactNativeReanimated.useDerivedValue)(() => {
    "worklet";

    const currentIndex = animationMaps.length - 1;
    let isAnyClosing = false;
    for (let i = 0; i < animationMaps.length; i++) {
      if (animationMaps[i].closing.value > 0) {
        isAnyClosing = true;
        break;
      }
    }
    return currentIndex - Number(isAnyClosing);
  });
  const focusedIndex = props.state.index;

  // StackContext value - for overlays via useStack()
  const stackContextValue = (0, _react.useMemo)(() => ({
    flags,
    routeKeys,
    routes: state.routes,
    descriptors: state.descriptors,
    scenes: scenes,
    focusedIndex,
    stackProgress,
    optimisticFocusedIndex
  }), [routeKeys, state.routes, state.descriptors, scenes, focusedIndex, stackProgress, optimisticFocusedIndex, flags]);

  // ManagedStack context value
  const lifecycleValue = (0, _react.useMemo)(() => ({
    routes: state.routes,
    focusedIndex,
    descriptors: state.descriptors,
    closingRouteKeysShared: closingRouteKeys.shared,
    activeScreensLimit,
    handleCloseRoute,
    scenes,
    shouldShowFloatOverlay,
    stackProgress,
    optimisticFocusedIndex
  }), [state.routes, state.descriptors, focusedIndex, closingRouteKeys.shared, activeScreensLimit, handleCloseRoute, scenes, shouldShowFloatOverlay, stackProgress, optimisticFocusedIndex]);
  return {
    ...lifecycleValue,
    stackContextValue
  };
}

/**
 * HOC that wraps component with ManagedStack provider AND StackContext.
 * Used by blank-stack which manages local route state for closing animations.
 * Generic over descriptor type - defaults to BaseStackDescriptor.
 */
function withManagedStack(Component) {
  return function ManagedStackProvider(props) {
    const navigatorKey = props.state.key;

    // Clean up history when navigator unmounts
    (0, _react.useEffect)(() => {
      return () => {
        _history.HistoryStore.clearNavigator(navigatorKey);
      };
    }, [navigatorKey]);
    const {
      stackContextValue,
      ...lifecycleValue
    } = useManagedStackValue(props);
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_useStack.StackContext.Provider, {
      value: stackContextValue,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ManagedStackContext.Provider, {
        value: lifecycleValue,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, {
          ...lifecycleValue
        })
      })
    });
  };
}
