// react-native-screen-transitions/lib/commonjs/shared/hooks/gestures/use-build-gestures.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useBuildGestures = void 0;
var _native = require("@react-navigation/native");
var _react = require("react");
var _reactNative = require("react-native");
var _reactNativeGestureHandler = require("react-native-gesture-handler");
var _reactNativeReanimated = require("react-native-reanimated");
var _constants = require("../../constants");
var _keys = require("../../providers/screen/keys.provider");
var _animation = require("../../stores/animation.store");
var _gesture = require("../../stores/gesture.store");
var _gesture2 = require("../../types/gesture.types");
var _startScreenTransition = require("../../utils/animation/start-screen-transition");
var _checkGestureActivation = require("../../utils/gesture/check-gesture-activation");
var _determineDismissal = require("../../utils/gesture/determine-dismissal");
var _mapGestureToProgress = require("../../utils/gesture/map-gesture-to-progress");
var _resetGestureValues = require("../../utils/gesture/reset-gesture-values");
var _velocity = require("../../utils/gesture/velocity");
var _useStableCallbackValue = _interopRequireDefault(require("../use-stable-callback-value"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const useBuildGestures = ({
  scrollConfig,
  ancestorContext
}) => {
  const dimensions = (0, _reactNative.useWindowDimensions)();
  const {
    current
  } = (0, _keys.useKeys)();
  const navState = current.navigation.getState();
  const isFirstScreen = navState.routes.findIndex(r => r.key === current.route.key) === 0;
  const initialTouch = (0, _reactNativeReanimated.useSharedValue)({
    x: 0,
    y: 0
  });
  const gestureOffsetState = (0, _reactNativeReanimated.useSharedValue)(_gesture2.GestureOffsetState.PENDING);

  // Ref for external gesture coordination (e.g., swipeable lists)
  const panGestureRef = (0, _react.useRef)(undefined);
  const gestureAnimationValues = _gesture.GestureStore.getRouteGestures(current.route.key);
  const animations = _animation.AnimationStore.getAll(current.route.key);
  const {
    gestureDirection = _constants.DEFAULT_GESTURE_DIRECTION,
    gestureVelocityImpact = _constants.GESTURE_VELOCITY_IMPACT,
    gestureDrivesProgress = _constants.DEFAULT_GESTURE_DRIVES_PROGRESS,
    gestureActivationArea = _constants.DEFAULT_GESTURE_ACTIVATION_AREA,
    gestureResponseDistance,
    transitionSpec
  } = current.options;
  const gestureEnabled = Boolean(isFirstScreen ? false : current.options.gestureEnabled);
  const directions = (0, _react.useMemo)(() => {
    const directionsArray = Array.isArray(gestureDirection) ? gestureDirection : [gestureDirection];
    const isBidirectional = directionsArray.includes("bidirectional");
    return {
      vertical: directionsArray.includes("vertical") || isBidirectional,
      verticalInverted: directionsArray.includes("vertical-inverted") || isBidirectional,
      horizontal: directionsArray.includes("horizontal") || isBidirectional,
      horizontalInverted: directionsArray.includes("horizontal-inverted") || isBidirectional
    };
  }, [gestureDirection]);
  const handleDismiss = (0, _react.useCallback)(() => {
    // If an ancestor navigator is already dismissing, skip this dismiss to
    // avoid racing with the ancestor
    if (ancestorContext?.gestureAnimationValues.isDismissing?.value) {
      return;
    }
    const state = current.navigation.getState();
    const routeStillPresent = state.routes.some(route => route.key === current.route.key);
    if (!routeStillPresent) {
      return;
    }
    current.navigation.dispatch({
      ..._native.StackActions.pop(),
      source: current.route.key,
      target: state.key
    });
  }, [current, ancestorContext]);
  const onTouchesDown = (0, _useStableCallbackValue.default)(e => {
    "worklet";

    const firstTouch = e.changedTouches[0];
    initialTouch.value = {
      x: firstTouch.x,
      y: firstTouch.y
    };
    gestureOffsetState.value = _gesture2.GestureOffsetState.PENDING;
  });
  const onTouchesMove = (0, _useStableCallbackValue.default)((e, manager) => {
    "worklet";

    // If an ancestor navigator is already dismissing via gesture, block new gestures here.
    if (ancestorContext?.gestureAnimationValues.isDismissing?.value) {
      gestureOffsetState.value = _gesture2.GestureOffsetState.FAILED;
      manager.fail();
      return;
    }
    const touch = e.changedTouches[0];
    const {
      isSwipingDown,
      isSwipingUp,
      isSwipingRight,
      isSwipingLeft
    } = (0, _checkGestureActivation.applyOffsetRules)({
      touch,
      directions,
      manager,
      dimensions,
      gestureOffsetState,
      initialTouch: initialTouch.value,
      activationArea: gestureActivationArea,
      responseDistance: gestureResponseDistance
    });
    if (gestureOffsetState.value === _gesture2.GestureOffsetState.FAILED) {
      manager.fail();
      return;
    }

    // Keep pending until thresholds are met; no eager activation.
    if (gestureAnimationValues.isDragging?.value) {
      manager.activate();
      return;
    }
    const maxScrollY = scrollConfig.value?.contentHeight ? scrollConfig.value.contentHeight - scrollConfig.value.layoutHeight : 0;
    const maxScrollX = scrollConfig.value?.contentWidth ? scrollConfig.value.contentWidth - scrollConfig.value.layoutWidth : 0;
    const recognizedDirection = isSwipingDown || isSwipingUp || isSwipingRight || isSwipingLeft;
    const scrollCfg = scrollConfig.value;
    let shouldActivate = false;
    let activatedDirection = null;
    if (recognizedDirection) {
      if (directions.vertical && isSwipingDown) {
        shouldActivate = scrollCfg ? scrollCfg.y <= 0 : true;
        if (shouldActivate) activatedDirection = "vertical";
      }
      if (directions.horizontal && isSwipingRight) {
        shouldActivate = scrollCfg ? scrollCfg.x <= 0 : true;
        if (shouldActivate) activatedDirection = "horizontal";
      }
      if (directions.verticalInverted && isSwipingUp) {
        shouldActivate = scrollCfg ? scrollCfg.y >= maxScrollY : true;
        if (shouldActivate) activatedDirection = "vertical-inverted";
      }
      if (directions.horizontalInverted && isSwipingLeft) {
        shouldActivate = scrollCfg ? scrollCfg.x >= maxScrollX : true;
        if (shouldActivate) activatedDirection = "horizontal-inverted";
      }
    }
    if (recognizedDirection && !shouldActivate) {
      manager.fail();
      return;
    }
    if (shouldActivate && gestureOffsetState.value === _gesture2.GestureOffsetState.PASSED && !gestureAnimationValues.isDismissing?.value) {
      gestureAnimationValues.direction.value = activatedDirection;
      manager.activate();
      return;
    }
  });
  const onStart = (0, _useStableCallbackValue.default)(() => {
    "worklet";

    gestureAnimationValues.isDragging.value = 1;
    gestureAnimationValues.isDismissing.value = 0;
  });
  const onUpdate = (0, _useStableCallbackValue.default)(event => {
    "worklet";

    let gestureProgress = 0;
    const {
      translationX,
      translationY
    } = event;
    const {
      width,
      height
    } = dimensions;
    gestureAnimationValues.x.value = translationX;
    gestureAnimationValues.y.value = translationY;
    gestureAnimationValues.normalizedX.value = Math.max(-1, Math.min(1, translationX / width));
    gestureAnimationValues.normalizedY.value = Math.max(-1, Math.min(1, translationY / height));
    let maxProgress = 0;
    const allowedDown = directions.vertical;
    const allowedUp = directions.verticalInverted;
    const allowedRight = directions.horizontal;
    const allowedLeft = directions.horizontalInverted;
    if (allowedRight && event.translationX > 0) {
      const currentProgress = (0, _mapGestureToProgress.mapGestureToProgress)(translationX, dimensions.width);
      maxProgress = Math.max(maxProgress, currentProgress);
    }
    if (allowedLeft && event.translationX < 0) {
      const currentProgress = (0, _mapGestureToProgress.mapGestureToProgress)(-translationX, dimensions.width);
      maxProgress = Math.max(maxProgress, currentProgress);
    }
    if (allowedDown && event.translationY > 0) {
      const currentProgress = (0, _mapGestureToProgress.mapGestureToProgress)(translationY, dimensions.height);
      maxProgress = Math.max(maxProgress, currentProgress);
    }
    if (allowedUp && event.translationY < 0) {
      const currentProgress = (0, _mapGestureToProgress.mapGestureToProgress)(-translationY, dimensions.height);
      maxProgress = Math.max(maxProgress, currentProgress);
    }
    gestureProgress = maxProgress;
    if (gestureDrivesProgress) {
      animations.progress.value = 1 - gestureProgress;
    }
  });
  const onEnd = (0, _useStableCallbackValue.default)(event => {
    "worklet";

    const {
      shouldDismiss
    } = (0, _determineDismissal.determineDismissal)({
      event,
      directions,
      dimensions,
      gestureVelocityImpact
    });
    const spec = shouldDismiss ? transitionSpec?.close : transitionSpec?.open;
    (0, _resetGestureValues.resetGestureValues)({
      spec,
      gestures: gestureAnimationValues,
      shouldDismiss,
      event,
      dimensions
    });
    const initialVelocity = _velocity.velocity.calculateProgressVelocity({
      animations,
      shouldDismiss,
      event,
      dimensions,
      directions
    });
    (0, _startScreenTransition.startScreenTransition)({
      target: shouldDismiss ? "close" : "open",
      onAnimationFinish: shouldDismiss ? handleDismiss : undefined,
      spec: transitionSpec,
      animations,
      initialVelocity
    });
  });

  // Memoize gestures to keep stable references - critical for RNGH
  // Child gestures reference ancestor's pan via requireExternalGestureToFail,
  // so the pan gesture MUST be stable or children will reference stale objects
  return (0, _react.useMemo)(() => {
    const panGesture = _reactNativeGestureHandler.Gesture.Pan().withRef(panGestureRef).enabled(gestureEnabled).manualActivation(true).onTouchesDown(onTouchesDown).onTouchesMove(onTouchesMove).onStart(onStart).onUpdate(onUpdate).onEnd(onEnd);

    // Native gesture setup depends on whether this screen has gestures
    let nativeGesture;
    if (gestureEnabled) {
      // This screen has gestures - set up normal pan/native relationship
      nativeGesture = _reactNativeGestureHandler.Gesture.Native().requireExternalGestureToFail(panGesture);
      panGesture.blocksExternalGesture(nativeGesture);
    } else {
      // This screen has no gestures
      // Find nearest ancestor with gestureEnabled=true (attached pan)
      let activePanAncestor = ancestorContext;
      while (activePanAncestor && !activePanAncestor.gestureEnabled) {
        activePanAncestor = activePanAncestor.ancestorContext;
      }
      if (activePanAncestor?.panGesture) {
        // Found an ancestor with enabled pan - wait for it
        nativeGesture = _reactNativeGestureHandler.Gesture.Native().requireExternalGestureToFail(activePanAncestor.panGesture);
      } else {
        // No ancestor with enabled pan - plain native
        nativeGesture = _reactNativeGestureHandler.Gesture.Native();
      }
    }
    return {
      panGesture,
      panGestureRef,
      nativeGesture,
      gestureAnimationValues
    };
  }, [gestureEnabled, onTouchesDown, onTouchesMove, onStart, onUpdate, onEnd, gestureAnimationValues, ancestorContext]);
};
exports.useBuildGestures = useBuildGestures;
