// react-native-screen-transitions/lib/commonjs/shared/hooks/gestures/use-scroll-registry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useScrollRegistry = void 0;
var _react = require("react");
var _reactNativeReanimated = require("react-native-reanimated");
var _gestures = require("../../providers/gestures.provider");
var _useStableCallback = _interopRequireDefault(require("../use-stable-callback"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/** biome-ignore-all lint/style/noNonNullAssertion: <Will always consume context from GestureProvider> */

const useScrollRegistry = props => {
  const context = (0, _gestures.useGestureContext)();
  const {
    scrollConfig,
    ancestorContext
  } = context;
  const ancestorScrollConfigs = (0, _react.useMemo)(() => {
    const configs = [];
    let current = ancestorContext;
    while (current) {
      if (current.scrollConfig) {
        configs.push(current.scrollConfig);
      }
      current = current.ancestorContext;
    }
    return configs;
  }, [ancestorContext]);
  const scrollHandler = (0, _reactNativeReanimated.useAnimatedScrollHandler)({
    onScroll: event => {
      props.onScroll?.(event);
      const updateScrollPosition = v => {
        "worklet";

        if (v === null) {
          return {
            x: event.contentOffset.x,
            y: event.contentOffset.y,
            contentHeight: 0,
            contentWidth: 0,
            layoutHeight: 0,
            layoutWidth: 0
          };
        }
        v.x = event.contentOffset.x;
        v.y = event.contentOffset.y;
        return v;
      };
      scrollConfig.modify(updateScrollPosition);

      // Sync to ALL ancestors, not just immediate parent
      for (const ancestorConfig of ancestorScrollConfigs) {
        ancestorConfig.modify(updateScrollPosition);
      }
    }
  });
  const onContentSizeChange = (0, _useStableCallback.default)((width, height) => {
    props.onContentSizeChange?.(width, height);
    const updateContentSize = v => {
      "worklet";

      if (v === null) {
        return {
          x: 0,
          y: 0,
          layoutHeight: 0,
          layoutWidth: 0,
          contentWidth: width,
          contentHeight: height
        };
      }
      v.contentWidth = width;
      v.contentHeight = height;
      return v;
    };
    scrollConfig.modify(updateContentSize);
    for (const ancestorConfig of ancestorScrollConfigs) {
      ancestorConfig.modify(updateContentSize);
    }
  });
  const onLayout = (0, _useStableCallback.default)(event => {
    props.onLayout?.(event);
    const {
      width,
      height
    } = event.nativeEvent.layout;
    const updateLayout = v => {
      "worklet";

      if (v === null) {
        return {
          x: 0,
          y: 0,
          contentHeight: 0,
          contentWidth: 0,
          layoutHeight: height,
          layoutWidth: width
        };
      }
      v.layoutHeight = height;
      v.layoutWidth = width;
      return v;
    };
    scrollConfig.modify(updateLayout);
    for (const ancestorConfig of ancestorScrollConfigs) {
      ancestorConfig.modify(updateLayout);
    }
  });
  return {
    scrollHandler,
    onContentSizeChange,
    onLayout
  };
};
exports.useScrollRegistry = useScrollRegistry;
