// react-native-screen-transitions/lib/commonjs/shared/components/integrations/masked-view.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MaskedView;
var _reactNative = require("react-native");
var _constants = require("../../constants");
var _createTransitionAwareComponent = require("../create-transition-aware-component");
var _jsxRuntime = require("react/jsx-runtime");
const TransitionView = (0, _createTransitionAwareComponent.createTransitionAwareComponent)(_reactNative.View);
let LazyMaskedView = _reactNative.View;
try {
  LazyMaskedView = require("@react-native-masked-view/masked-view").default;
} catch (_) {
  // noop
}
function MaskedView({
  children,
  style: userStyles = {}
}) {
  if (LazyMaskedView === _reactNative.View) {
    return children;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(LazyMaskedView, {
    style: styles.root
    // @ts-expect-error
    ,
    maskElement: /*#__PURE__*/(0, _jsxRuntime.jsx)(TransitionView, {
      styleId: _constants.MASK_STYLE_ID,
      style: styles.rootMask
    }),
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(TransitionView, {
      styleId: _constants.CONTAINER_STYLE_ID,
      style: [styles.rootContainer, userStyles],
      children: children
    })
  });
}
const styles = _reactNative.StyleSheet.create({
  root: {
    flex: 1
  },
  rootMask: {
    backgroundColor: "white"
  },
  rootContainer: {
    flex: 1
  }
});
